#' @title VARI3
#' @description Approach to VARIant-VARIant interaction through VARIable thresholds and hypothesis testing. VARI3 automates the selection and analysis of the most promising SNPs to identify epistasis.
#' @import parallel
#' @import readr
#' @import data.table
#' @import dplyr
#' @param bfile The bfile variable is required, the user needs to define the path and name of the Plink bfile (bed, bim and fam) without the extension, the three files must be named in the same way. For example genotype.bed, genotype.bim, genotype.fam the variable bfile would be (bfile = "genotype").
#' @param out The out variable determines the output directory where all files generated by VARI3 will be written.
#' @param primarylist prymarylist refers to the file with the list of genes where the selection of primary SNPs is searched, the file must not have a header.
#' @param genelist genelist refers to the file with the list of genes where you want to look for the epistasis, the secondary SNPs, the file must not have header.
#' @param plink The plink variable is by default "plink" but it is necessary to say the path of the previously downloaded plink file if it has not been added to the system executable path.
#' @param core The core variable is by default core = detectCores() -1. It refers to the number of cores used.
#' @param varpl The varpl variable is by default "annotate_variation.pl", but it is necessary to say the path of the .pl file if it has not been added to the system executable path.
#' @param ANNOVAR The ANNOVAR variable is by default ANNOVAR = T, for the annotation with ANNOVAR. Note, at the moment the functionality with false is not implemented.
#' @param clump The clump variable is by default clump = T, to perform clumping. We recommend clumping if the analysis is for more than 2000 people.
#' @param build The variable build is used to determine the version of the reference genome, it is by default (build = "hg19").
#' @param db The variable db is necessary to define where annovar reference databases are located, normally in the downloaded file they follow the path "annovar/humandb/".
#' @param Wu The variable Wu refers to the proposed correction for epistasis tests, by default it is Wu = T.
#' @param covar The covar variable refers to the file where the covariates are located. For example, (covar = "covarfile.cov"). By default it is null.
#' @param ncov The ncov variable refers to how many covariates are selected from the covariables file, they are selected depending on the column we want, as in plink. For example to select the first four of a file (ncov = "1-4")
#' @param AsT The AsT variable refers to the p.maximum value for selecting SNPs in the association analysis and is by default (AsT = 0.0001).
#' @param MAF The MAF variable refers to the MAF from which SNPs are selected for analysis, by default (MAF = 0.01) to avoid very rare SNPs.
#' @details This function has developed by Alejandro Cisterna Garcia as part of his PhD mentored by Juan Antonio Botía Blaya.
#' Help documentation has generated by roxygen2.
#' @export

# Función
VARI3 = function(bfile,
                        out = ".",
                        primarylist = NULL,
                        genelist = NULL,
                        plink = "plink",
                        core = detectCores() -1,
                        varpl = "annotate_variation.pl",
                        ANNOVAR = T,
                        clump = T,
                        build = "hg19",
                        db = "../../Doctorado/Annovar/annovar.latest/annovar/humandb/",
                          Wu = T,
                        covar = NULL,
                        ncov = "1-2",
                        AsT = 0.0001,
                        MAF = 0.01) {
  # No genlist
  if (is.null(genelist)){
    if (!is.null(covar)){
      # Logistic regresion
      system(paste0("",plink," --bfile ",bfile," --logistic --covar ",covar," --covar-number ",ncov," --out ",out,"/covariatesall --threads ",core,""))

      # Assoc analysis
      system(paste0("",plink," --bfile ",bfile," --assoc --out ",out,"/assocall --threads ",core,""))

      logistic = read_delim(paste0("",out,"/covariatesall.assoc.logistic"),
                            delim = " ", col_types = "ncnccnnnn")
      names(logistic) = c("CHR" ,"SNP", "BP", "A1", "TEST", "NMISS", "OR", "STAT", "P")
      logadd = filter(logistic, TEST == "       ADD")
      assoc = read_delim(paste0("",out,"/assocall.assoc"),
                         delim = " ", col_types = "ncncnncnnn")

      # Filter variants with treshold 10^-5
      assocrelevant = filter(logadd, logadd$P < AsT)

      write.table(assocrelevant , file = paste0("",out,"/assocrelevant.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)

      assocrelevant = read_delim(paste0("",out,"/assocrelevant.txt"),
                                 delim = " ", col_types = "ncnccnnnn")
      loginner = assocrelevant[c("SNP","OR","P")]
      names(loginner) = c("SNP" ,"ORl", "Pl")

      # Full_join by position assoc and logistic
      names(assoc)[2] = "SNP"
      assocrelevant = inner_join(assoc, loginner, by ="SNP")

      # Filter variants more frequently than 1% in our population in our case 4700 variants
      assocrelevant = filter(assocrelevant, assocrelevant$`     F_A` >= MAF | assocrelevant$`     F_U` >= MAF )

      if (clump){

        # Clumping analysis

        system(paste0("",plink," --bfile ",bfile," --clump ",out,"/assocall.assoc --out ",out,"/assocldclump --threads ",core,""))

        clumpdata = read_delim(paste0("",out,"/assocldclump.clumped"),
                               delim = " ", col_types = "nncnnnnnnnnc")

        # Mergue assocrelevant and clump
        aux = clumpdata[c(1, 2,3)]
        names(aux)[3] = "BP"
        names(assocrelevant)[2] = "BP"

        # Full_join by position, in our case 437 variants
        X = inner_join(assocrelevant, aux, by ="BP")

        # si no covs 11 y 12 si covs 13, 14
        assocclumpmergue = X[,-c(13,14)]
        if (nrow(assocclumpmergue) > 100){
          # Filter variants by frequency in the population
          t1 = filter(assocclumpmergue, assocclumpmergue$`     F_A` <= 0.05)
          t2 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.05 & assocclumpmergue$`     F_A` <= 0.1)
          t3 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.1 & assocclumpmergue$`     F_A` <= 0.2)
          t4 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.2 & assocclumpmergue$`     F_A` <= 0.3)
          t5 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.3 & assocclumpmergue$`     F_A` <= 0.4)
          t6 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.4 )

          # Sort a dataframe by p.value column
          t1ordenado = arrange(t1, t1$Pl)
          t2ordenado = arrange(t2, t2$Pl)
          t3ordenado = arrange(t3, t3$Pl)
          t4ordenado = arrange(t4, t4$Pl)
          t5ordenado = arrange(t5, t5$Pl)
          t6ordenado = arrange(t6, t6$Pl)

          # We select certain variants according to p value
          t1seleccion = t1ordenado[1:5,]
          t2seleccion = t2ordenado[1:5,]
          t3seleccion = t3ordenado[1:20,]
          t4seleccion = t4ordenado[1:20,]
          t5seleccion = t5ordenado[1:20,]
          t6seleccion = t6ordenado[1:30,]

          assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                       t4seleccion, t5seleccion, t6seleccion)
        } # Cierre if de clump si

        if (ANNOVAR){

          assocclumpmergue <- na.omit(assocclumpmergue)

          # Format for ANNOVAR
          names(assocclumpmergue) =c("CHR","SNP","BP","A1","F_A",
                                     "F_U","A2","CHISQ","P","OR","ORl","Pl")

          assocclumpmergue$SNPA = assocclumpmergue$SNP

          # Replace the chromosome to keep the position in the SNP column
          #for(ii in 22:1){
          #  assocclumpmergue$SNP = gsub(paste0("",ii,":"),
          #                              "", assocclumpmergue$SNP)
          #}

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)

          # He cambiado a BP por el problema de Bernabe
          asclumpano = data.frame(assocclumpmergue$CHR,assocclumpmergue$BP,
                                  assocclumpmergue$BP,assocclumpmergue$A2,
                                  assocclumpmergue$A1 , stringsAsFactors = F)

          # Format the file
          asclumpano$assocclumpmergue.CHR = as.numeric(asclumpano$assocclumpmergue.CHR)
          asclumpano$assocclumpmergue.BP = as.numeric(asclumpano$assocclumpmergue.BP)
          asclumpano$assocclumpmergue.BP.1 = as.numeric(asclumpano$assocclumpmergue.BP.1)
          asclumpano$assocclumpmergue.A2 = gsub("   ", "", asclumpano$assocclumpmergue.A2)
          asclumpano$assocclumpmergue.A1 = gsub("   ", "", asclumpano$assocclumpmergue.A1)

          # Write the file that is the ANNOVAR input
          write.table(asclumpano, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))



          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))

          if (identical(tx, !character(0))){
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
          }else{
          asevf = NULL
          }
          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))

          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

          if (length(asevf) != 0){
          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "BP"
          aux$BP = as.numeric(aux$BP)
          assocclumpmergue$BP = as.numeric(assocclumpmergue$BP)

          # Full_join by position
          X = full_join(assocclumpmergue, aux, by ="BP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "BP"
          aux2$BP = as.numeric(aux2$BP)
          X = full_join(X, aux2, by ="BP")
          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/aaclummerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "BP"
            aux2$BP = as.numeric(aux2$BP)
            assocclumpmergue$BP = as.numeric(assocclumpmergue$BP)
            aux2 = unique(aux2)

            X = full_join(assocclumpmergue, aux2, by ="BP")


            # # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)

          }


          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnnnncccc")

          # Make the set of SNPs to do epistasis analysis
          write.table(c("SET", asclumanovar$SNPA, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## ANNOVAR else de annovar no clump si
        else {

          names(assocclumpmergue) = c("CHR","SNP","BP","A1","F_A",
                                      "F_U","A2","CHISQ","P","OR",
                                      "ORl","Pl")

          write.table(assocclumpmergue, file = paste0("",out,"/assocclump.txt"),
                      sep = " ",row.names = F, col.names = F, quote = F)

          write.table(c("SET", assocclumpmergue$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ",row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        } ## cierre annovar no clump si
      } ## clump supuesto else de clump no (nuevo annovar)
      else {
        t1 = filter(assocrelevant, assocrelevant$`     F_A` <= 0.05)
        t2 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.05 & assocrelevant$`     F_A` <= 0.1)
        t3 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.1 & assocrelevant$`     F_A` <= 0.2)
        t4 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.2 & assocrelevant$`     F_A` <= 0.3)
        t5 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.3 & assocrelevant$`     F_A` <= 0.4)
        t6 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.4 )

        # Ordena un dataframe por una columna
        t1ordenado = arrange(t1, t1$Pl)
        t2ordenado = arrange(t2, t2$Pl)
        t3ordenado = arrange(t3, t3$Pl)
        t4ordenado = arrange(t4, t4$Pl)
        t5ordenado = arrange(t5, t5$Pl)
        t6ordenado = arrange(t6, t6$Pl)

        # Seleccionamos según filtrado
        t1seleccion = t1ordenado[1:5,]
        t2seleccion = t2ordenado[1:5,]
        t3seleccion = t3ordenado[1:20,]
        t4seleccion = t4ordenado[1:20,]
        t5seleccion = t5ordenado[1:20,]
        t6seleccion = t6ordenado[1:30,]

        seleccion = bind_rows(t1seleccion, t2seleccion, t3seleccion, t4seleccion,
                              t5seleccion, t6seleccion)

        if (ANNOVAR){
          # Preparamos para ANNOVAR
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR",
                               "ORl","Pl")
          seleccion$SNPA = seleccion$SNP



          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          assocanovar = data.frame(seleccion$CHR,seleccion$BP,seleccion$BP,
                                   seleccion$A2,seleccion$A1 , stringsAsFactors = F)

          # Format the file
          assocanovar$seleccion.CHR = as.numeric(assocanovar$seleccion.CHR)
          assocanovar$seleccion.BP = as.numeric(assocanovar$seleccion.BP)
          assocanovar$seleccion.BP.1 = as.numeric(assocanovar$seleccion.BP.1)
          assocanovar$seleccion.A2 = gsub("   ", "", assocanovar$seleccion.A2)
          assocanovar$seleccion.A1 = gsub("   ", "", assocanovar$seleccion.A1)

          # Write the file that is te ANNOVAR input
          write.table(assocanovar, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove quotes from the file
          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovar.txt"))


          # ANNOVAR command:
          # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))


          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove "SNV" from the file
          #system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)
          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))
          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "BP"
          aux$BP = as.numeric(aux$BP)
          seleccion$BP = as.numeric(seleccion$BP)

          # Full_join by position
          X = full_join(seleccion, aux, by ="BP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "BP"
          aux2$BP = as.numeric(aux2$BP)


          X = full_join(X, aux2, by ="BP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/assocannovarmerge.txt"),
                      sep = " ", row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/assocannovarmerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovarmerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovarmerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovarmerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/assocannovarmerge.txt"),
                                    delim = " ", col_types = "nnncnncnnnnncccc")

          write.table(c("SET", asclumanovar$SNPA, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## Cierre annovar si clump No
        else{

          names(seleccion) = c("CHR","SNP","BP","A1","F_A","F_U","A2","CHISQ","P","OR")

          write.table(seleccion, file = paste0("",out,"/assocfilter.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", seleccion$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        } # Cierre de ANNOVAR no EN CLUMP NO
      } ## Cierre de clump no
    } ## Cierre covs si
    else{

      system(paste0("",plink," --bfile ",bfile," --assoc --out ",out,"/assocall --threads ",core,""))

      assoc = read_delim(paste0("",out,"/assocall.assoc"),
                         delim = " ", col_types = "ncncnncnnn")

      assocrelevant = filter(assoc, assoc$`           P` < AsT)

      write.table(assocrelevant , file = paste0("",out,"/assocrelevant.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)

      assocrelevant = read_delim(paste0("",out,"/assocrelevant.txt"),
                                 delim = " ", col_types = "ncncnncnnn")

      assocrelevant = filter(assocrelevant, assocrelevant$`     F_A` >= MAF)

      if (clump){
        system(paste0("",plink," --bfile ",bfile," --clump ",out,"/assocall.assoc --out ",out,"/assocldclump --threads ",core,""))

        clumpdata = read_delim(paste0("",out,"/assocldclump.clumped"),
                               delim = " ", col_types = "nncnnnnnnnnc")

        # Filtrar assocrelevant para quedarnos solo las clump
        aux = clumpdata[c(1, 2,3)]
        names(aux)[3] = "SNP"
        names(assocrelevant)[2] = "SNP"

        # Full_join by position
        X = inner_join(assocrelevant, aux, by ="SNP")

        # si no covs 11 y 12 si covs 13, 14
        assocclumpmergue = X[,-c(11,12)]
        if (nrow(assocclumpmergue) > 100){
          t1 = filter(assocclumpmergue, assocclumpmergue$`     F_A` <= 0.05)
          t2 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.05 & assocclumpmergue$`     F_A` <= 0.1)
          t3 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.1 & assocclumpmergue$`     F_A` <= 0.2)
          t4 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.2 & assocclumpmergue$`     F_A` <= 0.3)
          t5 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.3 & assocclumpmergue$`     F_A` <= 0.4)
          t6 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.4 )

          # Ordena un dataframe por una columna
          t1ordenado = arrange(t1, t1$`           P`)
          t2ordenado = arrange(t2, t2$`           P`)
          t3ordenado = arrange(t3, t3$`           P`)
          t4ordenado = arrange(t4, t4$`           P`)
          t5ordenado = arrange(t5, t5$`           P`)
          t6ordenado = arrange(t6, t6$`           P`)

          # Seleccionamos según filtrado
          t1seleccion = t1ordenado[1:5,]
          t2seleccion = t2ordenado[1:5,]
          t3seleccion = t3ordenado[1:20,]
          t4seleccion = t4ordenado[1:20,]
          t5seleccion = t5ordenado[1:20,]
          t6seleccion = t6ordenado[1:30,]

          assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                       t4seleccion, t5seleccion, t6seleccion)
        } # Cierre del if del clump
        if (ANNOVAR){
          # Preparamos para ANNOVAR
          names(assocclumpmergue) = c("CHR","SNP","BP","A1","F_A",
                                      "F_U","A2","CHISQ","P","OR")
          assocclumpmergue$SNPA = assocclumpmergue$SNP


          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          asclumpano = data.frame(assocclumpmergue$CHR,assocclumpmergue$BP,
                                  assocclumpmergue$BP,assocclumpmergue$A2,
                                  assocclumpmergue$A1, stringsAsFactors = F)

          # Format the file
          asclumpano$assocclumpmergue.CHR = as.numeric(asclumpano$assocclumpmergue.CHR)
          asclumpano$assocclumpmergue.BP = as.numeric(asclumpano$assocclumpmergue.BP)
          asclumpano$assocclumpmergue.BP.1 = as.numeric(asclumpano$assocclumpmergue.BP.1)
          asclumpano$assocclumpmergue.A2 = gsub("   ", "", asclumpano$assocclumpmergue.A2)
          asclumpano$assocclumpmergue.A1 = gsub("   ", "", asclumpano$assocclumpmergue.A1)

          # Write the file that is te ANNOVAR input
          write.table(asclumpano, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove quotes from the file
          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovar.txt"))

          # ANNOVAR command:
          # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))

          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove "SNV" from the file
          #system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          # TIENE 0 EN ASEVF
          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          if (length(asevf) != 0){
          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
          }
          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))
          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

          #

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "BP"
            aux$BP = as.numeric(aux$BP)
            assocclumpmergue$BP = as.numeric(assocclumpmergue$BP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(assocclumpmergue, aux, by ="BP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "BP"
            aux2$BP = as.numeric(aux2$BP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="BP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"), sep = " ",
                        row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "BP"
            aux2$BP = as.numeric(aux2$BP)
            assocclumpmergue$BP = as.numeric(assocclumpmergue$BP)
            aux2 = unique(aux2)



            X = full_join(assocclumpmergue, aux2, by ="BP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }
          #



          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnncccc")

          write.table(c("SET", asclumanovar$SNPA, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)
          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## CIERRE ANNOVAR SI CLUMP SI
        else {
          names(assocclumpmergue) = c("CHR","SNP","BP","A1","F_A",
                                      "F_U","A2","CHISQ","P","OR")

          write.table(assocclumpmergue, file = paste0("",out,"/assocclump.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", assocclumpmergue$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        }## CIERRE ANNOVAR NO CLUMP SI
      }## cierre clump Si
      else {
        t1 = filter(assocrelevant, assocrelevant$`     F_A` <= 0.05)
        t2 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.05 & assocrelevant$`     F_A` <= 0.1)
        t3 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.1 & assocrelevant$`     F_A` <= 0.2)
        t4 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.2 & assocrelevant$`     F_A` <= 0.3)
        t5 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.3 & assocrelevant$`     F_A` <= 0.4)
        t6 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.4 )

        # Ordena un dataframe por una columna
        t1ordenado = arrange(t1, t1$`           P`)
        t2ordenado = arrange(t2, t2$`           P`)
        t3ordenado = arrange(t3, t3$`           P`)
        t4ordenado = arrange(t4, t4$`           P`)
        t5ordenado = arrange(t5, t5$`           P`)
        t6ordenado = arrange(t6, t6$`           P`)

        # Seleccionamos según filtrado
        t1seleccion = t1ordenado[1:5,]
        t2seleccion = t2ordenado[1:5,]
        t3seleccion = t3ordenado[1:20,]
        t4seleccion = t4ordenado[1:20,]
        t5seleccion = t5ordenado[1:20,]
        t6seleccion = t6ordenado[1:30,]

        seleccion = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                              t4seleccion, t5seleccion, t6seleccion)
        ## Apertura clump no annovar si
        if (ANNOVAR) {# Preparamos para ANNOVAR
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR")
          seleccion$SNPA = seleccion$SNP


          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          assocanovar = data.frame(seleccion$CHR,seleccion$BP,
                                   seleccion$BP,seleccion$A2,
                                   seleccion$A1, stringsAsFactors = F)

          # Format the file
          assocanovar$seleccion.CHR = as.numeric(assocanovar$seleccion.CHR)
          assocanovar$seleccion.BP = as.numeric(assocanovar$seleccion.BP)
          assocanovar$seleccion.BP.1 = as.numeric(assocanovar$seleccion.BP.1)
          assocanovar$seleccion.A2 = gsub("   ", "", assocanovar$seleccion.A2)
          assocanovar$seleccion.A1 = gsub("   ", "", assocanovar$seleccion.A1)

          # Write the file that is te ANNOVAR input
          write.table(assocanovar, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove quotes from the file
          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovar.txt"))


          # ANNOVAR command:
          # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))


          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove "SNV" from the file
          #system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)
          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))

          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "BP"
          aux$BP = as.numeric(aux$BP)
          seleccion$BP = as.numeric(seleccion$BP)

          # Full_join by position
          X = full_join(seleccion, aux, by ="BP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "BP"
          aux2$BP = as.numeric(aux2$BP)


          X = full_join(X, aux2, by ="BP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/assocannovarmerge.txt"),
                      sep = " ", row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/assocannovarmerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovarmerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovarmerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovarmerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/assocannovarmerge.txt"),
                                    delim = " ", col_types = "nnncnncnnncccc")

          write.table(c("SET", asclumanovar$SNPA, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## cIERRE ANNOVAR SI CLUMP NO
        else{
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR")
          write.table(seleccion, file = paste0("",out,"/assocfilter.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", seleccion$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        }## CIERRE ANNOVAR NO CLUMP NO
      } ## Cierre clump no
    } ## Cierre covs no

    # Generación de epistasis (Parametrizar los threads)


    if (Wu){

      # Generation of epistasis with WU joint effects correction (Same as CASSI, does not test for set so we have to use PLINK) Joint-effects applies correction that there are 5 per cell of the contingency table.
      system(paste0("",plink," --bfile ",bfile," --fast-epistasis joint-effects --set-test --set ",out,"/epistasisrset.set --out ",out,"/episelecrjoint -set-by-all --threads ",core,""))
    }
    else{
      # Without Joint effects
      system(paste0("",plink," --bfile ",bfile," --fast-epistasis --set-test --set ",out,"/epistasisrset.set --out ",out,"/episelecr -set-by-all --threads ",core,""))
    }


    tx  = readLines(paste0("",out,"/episelecrjoint.epi.cc.summary"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/episelecrjoint.epi.cc.summary"),
                sep = " ", row.names = F, col.names = F, quote = F)

    tx  = readLines(paste0("",out,"/episelecrjoint.epi.cc"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/episelecrjoint.epi.cc"),
                sep = " ", row.names = F, col.names = F, quote = F)

    if (ANNOVAR & !is.null(covar) & clump){
      # informe de epistasis final para ejecución estandar

      tx  = readLines(paste0("",out,"/assocall.assoc"))
      tx  = gsub(pattern = "\"", replace = "", x = tx)
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/assocall.assoc"),
                  sep = " ", row.names = F, col.names = F, quote = F)

      # Remove quotes and spaces
      #system(paste0("sed -i 's/\"//g' ",out,"/assocall.assoc"))
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocall.assoc"))

      # Read assoc data
      assocall100 = read_delim(paste0("",out,"/assocall.assoc"),
                               delim = " ", col_types = "ncncnncnnn")
      names(assocall100) = c("CHR" ,"SNP", "BP", "A1", "F_A", "F_U", "A2", "CHISQ", "P", "OR")

      # Read epistasis results
      episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                             delim = " ", col_types = "ncnnnnnc")
      names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
      episuc100 = mutate(episuc100, TBf = 0.05/episuc100$N_TOT)
      names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP", "TBf")

      # Read variants and annovar data
      assocrelevant100 = read_delim(paste0("",out,"/assocrelevant.txt"),
                                    delim = " ", col_types = "ncncnncnnn")

      # Bernabe
      asclumanovar100= read_delim(paste0("",out,"/aaclummerge.txt"),
                                  delim = " ", col_types = "ncncnncnnnnnccc")
      #asclumanovar100= read_delim(paste0("",out,"/aaclummerge.txt"),
      #                            delim = " ", col_types = "nnncnncnnncccccc")



      # Keep the epistasis SNP and mergue with assoc data

      aux = episuc100[8]
      names(aux) = "SNP"
      names(assocall100)[2] = "SNP"

      # Full_join by position
      X = inner_join(assocall100, aux, by ="SNP")

      X$SNPA = X$SNP

      # Replace the chromosome to keep the position in the SNP column
      # for(ii in 22:1){
      #  X$SNP = gsub(paste0("",ii,":"), "", X$SNP)
      #}

      # Format for ANNOVAR
      # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
      snpannovar = data.frame(X$CHR,X$BP,X$BP,X$A2,X$A1 , stringsAsFactors = F)

      # Format the file

      snpannovar$X.CHR = as.numeric(snpannovar$X.CHR)
      snpannovar$X.BP = as.numeric(snpannovar$X.BP)
      snpannovar$X.BP.1 = as.numeric(snpannovar$X.BP.1)
      snpannovar$X.A2 = gsub("   ", "", snpannovar$X.A2)
      snpannovar$X.A1 = gsub("   ", "", snpannovar$X.A1)

      # Write the file that is the ANNOVAR input
      write.table(snpannovar, file = paste0("",out,"/snpannovar.txt"),
                  sep = " ", row.names = F, col.names = F, quote = F)


      # Remove quotes from the file
      #system(paste0("sed -i 's/\"//g' ",out,"/snpannovar.txt"))

      # ANNOVAR command:
      # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/
      system(paste0("",varpl," -out ",out,"/snpannovar -build ",build," ",out,"/snpannovar.txt ",db,""))


      tx  = readLines(paste0("",out,"/snpannovar.exonic_variant_function"))
      tx  = gsub(pattern = "SNV", replace = "", x = tx)
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)


        write.table(tx, file = paste0("",out,"/snpannovar.exonic_variant_function"),
                    sep = " ", row.names = F, col.names = F, quote = F)


        # Remove "SNV" from the file
        #system(paste0("sed -i 's/SNV//g' ",out,"/snpannovar.exonic_variant_function"))

        # Remove the spaces and tabs to generate only simple spaces throughout the file
        #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.exonic_variant_function"))

        #  Load in R ANNOVAR file
        asevf = read_delim(paste0("",out,"/snpannovar.exonic_variant_function"),
                           delim = " ", col_types = "cccccccc", col_names = F)
        if (length(asevf) != 0){
        names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
        }
        # Revisar identical


      # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
      # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

      tx  = readLines(paste0("",out,"/snpannovar.variant_function"))
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/snpannovar.variant_function"),
                  sep = " ", row.names = F, col.names = F, quote = F)



      # Same for the other file
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.variant_function"))
      asvf = read_delim(paste0("",out,"/snpannovar.variant_function"),
                        delim = " ", col_types = "ccccccc", col_names = F)
      names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

      #
      if (length(asevf) != 0){
        # Mergue data frames
        aux = asevf[c("tipo", "ini")]
        names(aux)[2] = "BP"
        aux$BP = as.numeric(aux$BP)
        X$BP = as.numeric(X$BP)
        aux = unique(aux)

        # Full_join by position
        X = full_join(X, aux, by ="BP")

        # Same for the other file
        aux2 = asvf[c("loc","gen","ini")]
        names(aux2)[3] = "BP"
        aux2$BP = as.numeric(aux2$BP)
        aux2 = unique(aux2)

        X = full_join(X, aux2, by ="BP")

        # Write the file with PLINK and ANNOVAR data
        write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)
      }else{
        aux2 = asvf[c("loc","gen","ini")]
        names(aux2)[3] = "BP"
        aux2$BP = as.numeric(aux2$BP)
        X$BP = as.numeric(X$BP)
        aux2 = unique(aux2)



         X = full_join(X, aux2, by ="BP")




        # # Write the file with PLINK and ANNOVAR data
         write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                     sep = " ", row.names = F, col.names = T, quote = F)
      }

      tx  = readLines(paste0("",out,"/snpepistasisannovar.txt"))
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/snpepistasisannovar.txt"),
                  sep = " ", row.names = F, col.names = F, quote = F)

      #system(paste0("sed -i 's/\"//g' ",out,"/snpepistasisannovar.txt"))
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpepistasisannovar.txt"))

      # Bernab
      mergueprimarisnp = asclumanovar100[c(2,5,11,12,14,15)]
      #mergueprimarisnp = asclumanovar100[c(2,5,11,12,14,15,16)]
      #names(mergueprimarisnp) = c("SNP","F_A1","ORl1","Pl1","Tipo1","Loc1","Gen1")
      names(mergueprimarisnp) = c("SNP","F_A1","ORl1","Pl1","Loc1","Gen1")
      #mergueprimarisnp$SNP = as.numeric(mergueprimarisnp$SNP)

      episuc100$SNPA = episuc100[2]

      # Replace the chromosome to keep the position in the SNP column
      #for(ii in 22:1){
      #  episuc100$SNP = gsub(paste0("",ii,":"), "", episuc100$SNP)
      #}

      #episuc100$SNP = as.numeric(episuc100$SNP)


      # Full_join by position
      m1 = inner_join(episuc100, mergueprimarisnp, by ="SNP")

      if (length(asevf) != 0){
        merguesecundarysnp = X[c(2,5,9,10,12,13,14)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Tipo2", "Loc2", "Gen2")
      } else{
        merguesecundarysnp = X[c(2,5,9,10,12,13)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Loc2", "Gen2")
      }

      m1$SNPA2 = m1[8]

      # Replace the chromosome to keep the position in the SNP column
     # for(ii in 22:1){
    #    m1$BEST_SNP = gsub(paste0("",ii,":"), "", m1$BEST_SNP)
     # }

      #m1$BEST_SNP = as.numeric(m1$BEST_SNP)
      #merguesecundarysnp$BEST_SNP = as.numeric(merguesecundarysnp$BEST_SNP)

      m2 = inner_join(m1, merguesecundarysnp, by ="BEST_SNP")

      # Arreglar que en primer punto no hay asevf
      if (length(asevf) != 0){
        final = m2[c(10,6,16,15,22,11,12,13,14,17,18,19,20,21,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "TIPO2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,15,16,3:14)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)
      }else{
        final = m2[c(10,6,16,15,21,11,12,13,14,17,18,19,20,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "LOC2", "TBf")


        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        # Cambio para tipo bernabe
        epiinform = m5[c(1,2,14,15,3:13)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)

      }


    }

    if (ANNOVAR & is.null(covar) & clump){
      # informe de epistasis final para ejecución estandar

      tx  = readLines(paste0("",out,"/assocall.assoc"))
      tx  = gsub(pattern = "\"", replace = "", x = tx)
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/assocall.assoc"),
                  sep = " ", row.names = F, col.names = F, quote = F)

      # Remove quotes and spaces
      #system(paste0("sed -i 's/\"//g' ",out,"/assocall.assoc"))
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocall.assoc"))

      # Read assoc data
      assocall100 = read_delim(paste0("",out,"/assocall.assoc"),
                               delim = " ", col_types = "ncncnncnnn")
      names(assocall100) = c("CHR" ,"SNP", "BP", "A1", "F_A", "F_U", "A2", "CHISQ", "P", "OR")

      # Read epistasis results
      episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                             delim = " ", col_types = "ncnnnnnc")
      names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
      episuc100 = mutate(episuc100, TBf = 0.05/episuc100$N_TOT)
      names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP", "TBf")

      # Read variants and annovar data
      assocrelevant100 = read_delim(paste0("",out,"/assocrelevant.txt"),
                                    delim = " ", col_types = "ncncnncnnn")
      asclumanovar100= read_delim(paste0("",out,"/aaclummerge.txt"),
                                  delim = " ", col_types = "nnncnncnnncccccc")



      # Keep the epistasis SNP and mergue with assoc data

      aux = episuc100[8]
      names(aux) = "SNP"
      names(assocall100)[2] = "SNP"

      # Full_join by position
      X = inner_join(assocall100, aux, by ="SNP")

      X$SNPA = X$SNP

      # Replace the chromosome to keep the position in the SNP column
    #  for(ii in 22:1){
    #    X$SNP = gsub(paste0("",ii,":"), "", X$SNP)
     # }

      # Format for ANNOVAR
      # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
      snpannovar = data.frame(X$CHR,X$BP,X$BP,X$A2,X$A1 , stringsAsFactors = F)

      # Format the file

      snpannovar$X.CHR = as.numeric(snpannovar$X.CHR)
      snpannovar$X.BP = as.numeric(snpannovar$X.BP)
      snpannovar$X.BP.1 = as.numeric(snpannovar$X.BP.1)
      snpannovar$X.A2 = gsub("   ", "", snpannovar$X.A2)
      snpannovar$X.A1 = gsub("   ", "", snpannovar$X.A1)

      # Write the file that is the ANNOVAR input
      write.table(snpannovar, file = paste0("",out,"/snpannovar.txt"),
                  sep = " ", row.names = F, col.names = F, quote = F)

      # me he quedado aqui REVISANDO

      # Remove quotes from the file
      #system(paste0("sed -i 's/\"//g' ",out,"/snpannovar.txt"))

      # ANNOVAR command:
      # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/
      system(paste0("",varpl," -out ",out,"/snpannovar -build ",build," ",out,"/snpannovar.txt ",db,""))


      tx  = readLines(paste0("",out,"/snpannovar.exonic_variant_function"))
      tx  = gsub(pattern = "SNV", replace = "", x = tx)
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)

        write.table(tx, file = paste0("",out,"/snpannovar.exonic_variant_function"),
                    sep = " ", row.names = F, col.names = F, quote = F)


        # Remove "SNV" from the file
        #system(paste0("sed -i 's/SNV//g' ",out,"/snpannovar.exonic_variant_function"))

        # Remove the spaces and tabs to generate only simple spaces throughout the file
        #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.exonic_variant_function"))

        #  Load in R ANNOVAR file
        asevf = read_delim(paste0("",out,"/snpannovar.exonic_variant_function"),
                           delim = " ", col_types = "cccccccc", col_names = F)
        if (length(asevf) != 0){
        names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
      }

      # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
      # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

      tx  = readLines(paste0("",out,"/snpannovar.variant_function"))
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/snpannovar.variant_function"),
                  sep = " ", row.names = F, col.names = F, quote = F)



      # Same for the other file
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.variant_function"))
      asvf = read_delim(paste0("",out,"/snpannovar.variant_function"),
                        delim = " ", col_types = "ccccccc", col_names = F)
      names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

      #
      if (length(asevf) != 0){
        # Mergue data frames
        aux = asevf[c("tipo", "ini")]
        names(aux)[2] = "SNP"
        aux$SNP = as.numeric(aux$SNP)
        X$SNP = as.numeric(X$SNP)
        aux = unique(aux)

        # Full_join by position
        X = full_join(X, aux, by ="SNP")

        # Same for the other file
        aux2 = asvf[c("loc","gen","ini")]
        names(aux2)[3] = "SNP"
        aux2$SNP = as.numeric(aux2$SNP)
        aux2 = unique(aux2)

        X = full_join(X, aux2, by ="SNP")

        # Write the file with PLINK and ANNOVAR data
        write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)
      }else{
        aux2 = asvf[c("loc","gen","ini")]
        names(aux2)[3] = "BP"
        aux2$BP = as.numeric(aux2$BP)
        X$BP = as.numeric(X$BP)
        aux2 = unique(aux2)



        X = full_join(X, aux2, by ="BP")




        # Write the file with PLINK and ANNOVAR data
        write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)
      }

      tx  = readLines(paste0("",out,"/snpepistasisannovar.txt"))
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/snpepistasisannovar.txt"),
                  sep = " ", row.names = F, col.names = F, quote = F)

      #system(paste0("sed -i 's/\"//g' ",out,"/snpepistasisannovar.txt"))
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpepistasisannovar.txt"))


      # ESTO ES LO NORMAL mergueprimarisnp = asclumanovar100[c(2,5,10,9,12,13,14)]
      if (length(asevf) != 0){
      mergueprimarisnp = asclumanovar100[c(2,5,10,9,12,13,14)]
      names(mergueprimarisnp) = c("SNP","F_A1","OR1","P1","Tipo1","Loc1","Gen1")
      } else{ # IREM
      mergueprimarisnp = asclumanovar100[c(11,5,10,9,12,13)]
      names(mergueprimarisnp) = c("SNP","F_A1","OR1","P1","Loc1","Gen1")
      }
      #mergueprimarisnp$SNP = as.numeric(mergueprimarisnp$SNP)

      episuc100$SNPA = episuc100[2]

      # Replace the chromosome to keep the position in the SNP column
     # for(ii in 22:1){
     #    episuc100$SNP = gsub(paste0("",ii,":"), "", episuc100$SNP)
      #}

      #episuc100$SNP = as.numeric(episuc100$SNP)


      # Full_join by position
      m1 = inner_join(episuc100, mergueprimarisnp, by ="SNP")

      if (length(asevf) != 0){
        merguesecundarysnp = X[c(2,5,9,10,12,13,14)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","P2", "OR2", "Tipo2", "Loc2", "Gen2")
      } else{
        merguesecundarysnp = X[c(2,5,9,10,12,13)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","P2", "OR2", "Loc2", "Gen2")
      }

      m1$SNPA2 = m1[8]

      # Replace the chromosome to keep the position in the SNP column
     # for(ii in 22:1){
      #  m1$BEST_SNP = gsub(paste0("",ii,":"), "", m1$BEST_SNP)
      #}

      #m1$BEST_SNP = as.numeric(m1$BEST_SNP)
      #merguesecundarysnp$BEST_SNP = as.numeric(merguesecundarysnp$BEST_SNP)

      m2 = inner_join(m1, merguesecundarysnp, by ="BEST_SNP")

      if (length(asevf) != 0){
        final = m2[c(10,6,17,16,23,11,12,13,14,15,18,19,20,21,22,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "OR1", "P1", "TIPO1", "LOC1",
                         "F_A2", "P2", "OR2", "TIPO2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,16,17,3:15)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)
      }else{
        final = m2[c(2,6,16,15,21,11,12,13,14,17,18,19,20,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "OR1", "P1", "LOC1",
                         "F_A2", "P2", "OR2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,14,15,3:13)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)

      }


    }

    # Remove quotes and spaces
    #system(paste0("sed -i 's/\"//g' ",out,"/episelecrjoint.epi.cc.summary"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/episelecrjoint.epi.cc.summary"))
    #system(paste0("sed -i 's/\"//g' ",out,"/episelecrjoint.epi.cc"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/episelecrjoint.epi.cc"))

    if (ANNOVAR & !is.null(covar) & !clump){

      tx  = readLines(paste0("",out,"/assocall.assoc"))
      tx  = gsub(pattern = "\"", replace = "", x = tx)
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/assocall.assoc"),
                  sep = " ", row.names = F, col.names = F, quote = F)

      # Remove quotes and spaces
      #system(paste0("sed -i 's/\"//g' ",out,"/assocall.assoc"))
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocall.assoc"))

      # Read assoc data
      assocall100 = read_delim(paste0("",out,"/assocall.assoc"),
                               delim = " ", col_types = "ncncnncnnn")
      names(assocall100) = c("CHR" ,"SNP", "BP", "A1", "F_A", "F_U", "A2", "CHISQ", "P", "OR")

      # Read epistasis results
      episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                             delim = " ", col_types = "ncnnnnnc")
      names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
      episuc100 = mutate(episuc100, TBf = 0.05/episuc100$N_TOT)
      names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP", "TBf")

      # Read variants and annovar data
      assocrelevant100 = read_delim(paste0("",out,"/assocrelevant.txt"),
                                    delim = " ", col_types = "ncncnncnnn")
      asclumanovar100= read_delim(paste0("",out,"/assocannovarmerge.txt"),
                                  delim = " ", col_types = "nnncnncnnncccccc")



      # Keep the epistasis SNP and mergue with assoc data

      aux = episuc100[8]
      names(aux) = "SNP"
      names(assocall100)[2] = "SNP"

      # Full_join by position
      X = inner_join(assocall100, aux, by ="SNP")

      X$SNPA = X$SNP

      # Replace the chromosome to keep the position in the SNP column
      for(ii in 22:1){
        X$SNP = gsub(paste0("",ii,":"), "", X$SNP)
      }

      # Format for ANNOVAR
      # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
      snpannovar = data.frame(X$CHR,X$SNP,X$SNP,X$A2,X$A1 , stringsAsFactors = F)

      # Format the file

      snpannovar$X.CHR = as.numeric(snpannovar$X.CHR)
      snpannovar$X.SNP = as.numeric(snpannovar$X.SNP)
      snpannovar$X.SNP.1 = as.numeric(snpannovar$X.SNP.1)
      snpannovar$X.A2 = gsub("   ", "", snpannovar$X.A2)
      snpannovar$X.A1 = gsub("   ", "", snpannovar$X.A1)

      # Write the file that is the ANNOVAR input
      write.table(snpannovar, file = paste0("",out,"/snpannovar.txt"),
                  sep = " ", row.names = F, col.names = F, quote = F)


      # Remove quotes from the file
      #system(paste0("sed -i 's/\"//g' ",out,"/snpannovar.txt"))

      # ANNOVAR command:
      # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/
      system(paste0("",varpl," -out ",out,"/snpannovar -build ",build," ",out,"/snpannovar.txt ",db,""))


      tx  = readLines(paste0("",out,"/snpannovar.exonic_variant_function"))
      tx  = gsub(pattern = "SNV", replace = "", x = tx)
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      if (identical(tx, !character(0))){
        write.table(tx, file = paste0("",out,"/snpannovar.exonic_variant_function"),
                    sep = " ", row.names = F, col.names = F, quote = F)


        # Remove "SNV" from the file
        #system(paste0("sed -i 's/SNV//g' ",out,"/snpannovar.exonic_variant_function"))

        # Remove the spaces and tabs to generate only simple spaces throughout the file
        #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.exonic_variant_function"))

        #  Load in R ANNOVAR file
        asevf = read_delim(paste0("",out,"/snpannovar.exonic_variant_function"),
                           delim = " ", col_types = "cccccccc", col_names = F)
        names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
      }

      # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
      # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

      tx  = readLines(paste0("",out,"/snpannovar.variant_function"))
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/snpannovar.variant_function"),
                  sep = " ", row.names = F, col.names = F, quote = F)



      # Same for the other file
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.variant_function"))
      asvf = read_delim(paste0("",out,"/snpannovar.variant_function"),
                        delim = " ", col_types = "ccccccc", col_names = F)
      names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


      if (length(asevf) != 0){
        #
        # Mergue data frames
        aux = asevf[c("tipo", "ini")]
        names(aux)[2] = "SNP"
        aux$SNP = as.numeric(aux$SNP)
        X$SNP = as.numeric(X$SNP)
        aux = unique(aux)

        # Full_join by position
        X = full_join(X, aux, by ="SNP")

        # Same for the other file
        aux2 = asvf[c("loc","gen","ini")]
        names(aux2)[3] = "SNP"
        aux2$SNP = as.numeric(aux2$SNP)
        aux2 = unique(aux2)



        X = full_join(X, aux2, by ="SNP")

        # Write the file with PLINK and ANNOVAR data
        write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)
      }else{
        aux2 = asvf[c("loc","gen","ini")]
        names(aux2)[3] = "SNP"
        aux2$SNP = as.numeric(aux2$SNP)
        X$SNP = as.numeric(X$SNP)
        aux2 = unique(aux2)



        X = full_join(X, aux2, by ="SNP")




        # Write the file with PLINK and ANNOVAR data
        write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)
      }

      tx  = readLines(paste0("",out,"/snpepistasisannovar.txt"))
      tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
      write.table(tx, file = paste0("",out,"/snpepistasisannovar.txt"),
                  sep = " ", row.names = F, col.names = F, quote = F)

      #system(paste0("sed -i 's/\"//g' ",out,"/snpepistasisannovar.txt"))
      #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpepistasisannovar.txt"))

      mergueprimarisnp = asclumanovar100[c(2,5,11,12,14,15,16)]
      names(mergueprimarisnp) = c("SNP","F_A1","ORl1","Pl1","Tipo1","Loc1","Gen1")

      mergueprimarisnp$SNP = as.numeric(mergueprimarisnp$SNP)

      episuc100$SNPA = episuc100[2]

      # Replace the chromosome to keep the position in the SNP column
      for(ii in 22:1){
        episuc100$SNP = gsub(paste0("",ii,":"), "", episuc100$SNP)
      }

      episuc100$SNP = as.numeric(episuc100$SNP)


      # Full_join by position
      m1 = inner_join(episuc100, mergueprimarisnp, by ="SNP")


      if (length(asevf) != 0){
        merguesecundarysnp = X[c(2,5,9,10,12,13,14)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Tipo2", "Loc2", "Gen2")
      } else{
        merguesecundarysnp = X[c(2,5,9,10,12,13)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Loc2", "Gen2")
      }

      m1$SNPA2 = m1[8]

      # Replace the chromosome to keep the position in the SNP column
      for(ii in 22:1){
        m1$BEST_SNP = gsub(paste0("",ii,":"), "", m1$BEST_SNP)
      }

      m1$BEST_SNP = as.numeric(m1$BEST_SNP)
      merguesecundarysnp$BEST_SNP = as.numeric(merguesecundarysnp$BEST_SNP)
      merguesecundarysnp = unique(merguesecundarysnp)

      m2 = inner_join(m1, merguesecundarysnp, by ="BEST_SNP")

      if (length(asevf) != 0){
        final = m2[c(10,6,17,16,23,11,12,13,14,15,18,19,20,21,22,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "TIPO1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "TIPO2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,16,17,3:15)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)
      }else{
        final = m2[c(10,6,17,16,22,11,12,13,15,18,19,20,21,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,14,15,3:13)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)

      }


    }




    #system(paste0("sed -i 's/\"//g' ",out,"/epiinform.txt"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/epiinform.txt"))
  } # No genlist


  # Genelist no primary
  if(!is.null(genelist) & is.null(primarylist)){ # Genelist no es nulo y primary si
    # Always ANNOVAR T
    ANNOVAR = T

    # Read genelist
    genes <- read_tsv(paste0("",genelist,""), col_names = F)
    names(genes) <- "GEN"

    if (!is.null(covar)){

      # Logistic regresion
      system(paste0("",plink," --bfile ",bfile," --logistic --covar ",covar," --covar-number ",ncov," --out ",out,"/covariatesall --threads ",core,""))

      # Assoc analysis
      system(paste0("",plink," --bfile ",bfile," --assoc --out ",out,"/assocall --threads ",core,""))

      logistic = read_delim(paste0("",out,"/covariatesall.assoc.logistic"),
                            delim = " ", col_types = "ncnccnnnn")
      names(logistic) = c("CHR" ,"SNP", "BP", "A1", "TEST", "NMISS", "OR", "STAT", "P")
      logadd = filter(logistic, TEST == "       ADD")
      assoc = read_delim(paste0("",out,"/assocall.assoc"),
                         delim = " ", col_types = "ncncnncnnn")

      # Filter variants with treshold 10^-5
      assocrelevant = filter(logadd, logadd$P < AsT)

      write.table(assocrelevant , file = paste0("",out,"/assocrelevant.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)

      assocrelevant = read_delim(paste0("",out,"/assocrelevant.txt"),
                                 delim = " ", col_types = "ncnccnnnn")
      loginner = assocrelevant[c("SNP","OR","P")]
      names(loginner) = c("SNP" ,"ORl", "Pl")

      # Full_join by position assoc and logistic
      names(assoc)[2] = "SNP"
      assocrelevant = inner_join(assoc, loginner, by ="SNP")

      # Filter variants more frequently than 1% in our population in our case 4700 variants
      assocrelevant = filter(assocrelevant, assocrelevant$`     F_A` >= MAF | assocrelevant$`     F_U` >= MAF )

      if (clump){

        # Clumping analysis

        system(paste0("",plink," --bfile ",bfile," --clump ",out,"/assocall.assoc --out ",out,"/assocldclump --threads ",core,""))

        clumpdata = read_delim(paste0("",out,"/assocldclump.clumped"),
                               delim = " ", col_types = "nncnnnnnnnnc")

        # Mergue assocrelevant and clump
        aux = clumpdata[c(1, 2,3)]
        names(aux)[3] = "SNP"
        names(assocrelevant)[2] = "SNP"

        # Full_join by position, in our case 437 variants
        X = inner_join(assocrelevant, aux, by ="SNP")

        # si no covs 11 y 12 si covs 13, 14
        assocclumpmergue = X[,-c(13,14)]
        if (nrow(assocclumpmergue) > 100){
          # Filter variants by frequency in the population
          t1 = filter(assocclumpmergue, assocclumpmergue$`     F_A` <= 0.05)
          t2 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.05 & assocclumpmergue$`     F_A` <= 0.1)
          t3 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.1 & assocclumpmergue$`     F_A` <= 0.2)
          t4 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.2 & assocclumpmergue$`     F_A` <= 0.3)
          t5 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.3 & assocclumpmergue$`     F_A` <= 0.4)
          t6 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.4 )

          # Sort a dataframe by p.value column
          t1ordenado = arrange(t1, t1$Pl)
          t2ordenado = arrange(t2, t2$Pl)
          t3ordenado = arrange(t3, t3$Pl)
          t4ordenado = arrange(t4, t4$Pl)
          t5ordenado = arrange(t5, t5$Pl)
          t6ordenado = arrange(t6, t6$Pl)

          # We select certain variants according to p value
          t1seleccion = t1ordenado[1:5,]
          t2seleccion = t2ordenado[1:5,]
          t3seleccion = t3ordenado[1:20,]
          t4seleccion = t4ordenado[1:20,]
          t5seleccion = t5ordenado[1:20,]
          t6seleccion = t6ordenado[1:30,]

          assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                       t4seleccion, t5seleccion, t6seleccion)
        } # Cierre if de clump si

        if (ANNOVAR){

          # Format for ANNOVAR
          names(assocclumpmergue) =c("CHR","SNP","BP","A1","F_A",
                                     "F_U","A2","CHISQ","P","OR","ORl","Pl")

          assocclumpmergue$SNPA = assocclumpmergue$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assocclumpmergue$SNP = gsub(paste0("",ii,":"),
                                        "", assocclumpmergue$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          asclumpano = data.frame(assocclumpmergue$CHR,assocclumpmergue$SNP,
                                  assocclumpmergue$SNP,assocclumpmergue$A2,
                                  assocclumpmergue$A1 , stringsAsFactors = F)

          # Format the file
          asclumpano$assocclumpmergue.CHR = as.numeric(asclumpano$assocclumpmergue.CHR)
          asclumpano$assocclumpmergue.SNP = as.numeric(asclumpano$assocclumpmergue.SNP)
          asclumpano$assocclumpmergue.SNP.1 = as.numeric(asclumpano$assocclumpmergue.SNP.1)
          asclumpano$assocclumpmergue.A2 = gsub("   ", "", asclumpano$assocclumpmergue.A2)
          asclumpano$assocclumpmergue.A1 = gsub("   ", "", asclumpano$assocclumpmergue.A1)

          # Write the file that is the ANNOVAR input
          write.table(asclumpano, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))



          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))

          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

          # mequedo
          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(assocclumpmergue, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"), sep = " ",
                        row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux2 = unique(aux2)



            X = full_join(assocclumpmergue, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }

          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnnnncccc")
          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          # Filtra las variantes de la lista de genes
          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## ANNOVAR else de annovar no clump si
        else {

          names(assocclumpmergue) = c("CHR","SNP","BP","A1","F_A",
                                      "F_U","A2","CHISQ","P","OR",
                                      "ORl","Pl")

          write.table(assocclumpmergue, file = paste0("",out,"/assocclump.txt"),
                      sep = " ",row.names = F, col.names = F, quote = F)

          write.table(c("SET", assocclumpmergue$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ",row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        } ## cierre annovar no clump si
      } ## clump supuesto else de clump no (nuevo annovar)
      else {
        t1 = filter(assocrelevant, assocrelevant$`     F_A` <= 0.05)
        t2 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.05 & assocrelevant$`     F_A` <= 0.1)
        t3 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.1 & assocrelevant$`     F_A` <= 0.2)
        t4 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.2 & assocrelevant$`     F_A` <= 0.3)
        t5 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.3 & assocrelevant$`     F_A` <= 0.4)
        t6 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.4 )

        # Ordena un dataframe por una columna
        t1ordenado = arrange(t1, t1$Pl)
        t2ordenado = arrange(t2, t2$Pl)
        t3ordenado = arrange(t3, t3$Pl)
        t4ordenado = arrange(t4, t4$Pl)
        t5ordenado = arrange(t5, t5$Pl)
        t6ordenado = arrange(t6, t6$Pl)

        # Seleccionamos según filtrado
        t1seleccion = t1ordenado[1:5,]
        t2seleccion = t2ordenado[1:5,]
        t3seleccion = t3ordenado[1:20,]
        t4seleccion = t4ordenado[1:20,]
        t5seleccion = t5ordenado[1:20,]
        t6seleccion = t6ordenado[1:30,]

        seleccion = bind_rows(t1seleccion, t2seleccion, t3seleccion, t4seleccion,
                              t5seleccion, t6seleccion)

        if (ANNOVAR){
          # Preparamos para ANNOVAR
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR",
                               "ORl","Pl")
          seleccion$SNPA = seleccion$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            seleccion$SNP = gsub(paste0("",ii,":"), "", seleccion$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          assocanovar = data.frame(seleccion$CHR,seleccion$SNP,seleccion$SNP,
                                   seleccion$A2,seleccion$A1 , stringsAsFactors = F)

          # Format the file
          assocanovar$seleccion.CHR = as.numeric(assocanovar$seleccion.CHR)
          assocanovar$seleccion.SNP = as.numeric(assocanovar$seleccion.SNP)
          assocanovar$seleccion.SNP.1 = as.numeric(assocanovar$seleccion.SNP.1)
          assocanovar$seleccion.A2 = gsub("   ", "", assocanovar$seleccion.A2)
          assocanovar$seleccion.A1 = gsub("   ", "", assocanovar$seleccion.A1)

          # Write the file that is te ANNOVAR input
          write.table(assocanovar, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove quotes from the file
          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovar.txt"))


          # ANNOVAR command:
          # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))


          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove "SNV" from the file
          #system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)
          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))
          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(seleccion, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux2 = unique(aux2)



            X = full_join(seleccion, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }

          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnnnncccc")
          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          # Filtra las variantes de la lista de genes
          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
          # clumpnoglt

        } ## Cierre annovar si clump No
        else{

          names(seleccion) = c("CHR","SNP","BP","A1","F_A","F_U","A2","CHISQ","P","OR")

          write.table(seleccion, file = paste0("",out,"/assocfilter.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", seleccion$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        } # Cierre de ANNOVAR no EN CLUMP NO
      } ## Cierre de clump no
    } ## Cierre covs si
    else{

      system(paste0("",plink," --bfile ",bfile," --assoc --out ",out,"/assocall --threads ",core,""))

      assoc = read_delim(paste0("",out,"/assocall.assoc"),
                         delim = " ", col_types = "ncncnncnnn")

      assocrelevant = filter(assoc, assoc$`           P` < AsT)

      write.table(assocrelevant , file = paste0("",out,"/assocrelevant.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)

      assocrelevant = read_delim(paste0("",out,"/assocrelevant.txt"),
                                 delim = " ", col_types = "ncncnncnnn")

      assocrelevant = filter(assocrelevant, assocrelevant$`     F_A` >= MAF)

      if (clump){
        system(paste0("",plink," --bfile ",bfile," --clump ",out,"/assocall.assoc --out ",out,"/assocldclump --threads ",core,""))

        clumpdata = read_delim(paste0("",out,"/assocldclump.clumped"),
                               delim = " ", col_types = "nncnnnnnnnnc")

        # Filtrar assocrelevant para quedarnos solo las clump
        aux = clumpdata[c(1, 2,3)]
        names(aux)[3] = "SNP"
        names(assocrelevant)[2] = "SNP"

        # Full_join by position
        X = inner_join(assocrelevant, aux, by ="SNP")

        # si no covs 11 y 12 si covs 13, 14
        assocclumpmergue = X[,-c(11,12)]
        if (nrow(assocclumpmergue) > 100){
          t1 = filter(assocclumpmergue, assocclumpmergue$`     F_A` <= 0.05)
          t2 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.05 & assocclumpmergue$`     F_A` <= 0.1)
          t3 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.1 & assocclumpmergue$`     F_A` <= 0.2)
          t4 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.2 & assocclumpmergue$`     F_A` <= 0.3)
          t5 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.3 & assocclumpmergue$`     F_A` <= 0.4)
          t6 = filter(assocclumpmergue, assocclumpmergue$`     F_A` >= 0.4 )

          # Ordena un dataframe por una columna
          t1ordenado = arrange(t1, t1$`           P`)
          t2ordenado = arrange(t2, t2$`           P`)
          t3ordenado = arrange(t3, t3$`           P`)
          t4ordenado = arrange(t4, t4$`           P`)
          t5ordenado = arrange(t5, t5$`           P`)
          t6ordenado = arrange(t6, t6$`           P`)

          # Seleccionamos según filtrado
          t1seleccion = t1ordenado[1:5,]
          t2seleccion = t2ordenado[1:5,]
          t3seleccion = t3ordenado[1:20,]
          t4seleccion = t4ordenado[1:20,]
          t5seleccion = t5ordenado[1:20,]
          t6seleccion = t6ordenado[1:30,]

          assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                       t4seleccion, t5seleccion, t6seleccion)
        } # Cierre del if del clump
        if (ANNOVAR){
          # Format for ANNOVAR
          names(assocclumpmergue) =c("CHR","SNP","BP","A1","F_A",
                                     "F_U","A2","CHISQ","P","OR")

          assocclumpmergue$SNPA = assocclumpmergue$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assocclumpmergue$SNP = gsub(paste0("",ii,":"),
                                        "", assocclumpmergue$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          asclumpano = data.frame(assocclumpmergue$CHR,assocclumpmergue$SNP,
                                  assocclumpmergue$SNP,assocclumpmergue$A2,
                                  assocclumpmergue$A1 , stringsAsFactors = F)

          # Format the file
          asclumpano$assocclumpmergue.CHR = as.numeric(asclumpano$assocclumpmergue.CHR)
          asclumpano$assocclumpmergue.SNP = as.numeric(asclumpano$assocclumpmergue.SNP)
          asclumpano$assocclumpmergue.SNP.1 = as.numeric(asclumpano$assocclumpmergue.SNP.1)
          asclumpano$assocclumpmergue.A2 = gsub("   ", "", asclumpano$assocclumpmergue.A2)
          asclumpano$assocclumpmergue.A1 = gsub("   ", "", asclumpano$assocclumpmergue.A1)

          # Write the file that is the ANNOVAR input
          write.table(asclumpano, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))



          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          if (length(asevf) != 0){
            names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
          }
          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))

          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

          # mequedo
          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(assocclumpmergue, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"), sep = " ",
                        row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux2 = unique(aux2)



            X = full_join(assocclumpmergue, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }

          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnnnncccc")
          asclumanovar$SNPA = assocclumpmergue$SNPA

          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          # Filtra las variantes de la lista de genes
          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## CIERRE ANNOVAR SI CLUMP SI
        else {
          names(assocclumpmergue) = c("CHR","SNP","BP","A1","F_A",
                                      "F_U","A2","CHISQ","P","OR")

          write.table(assocclumpmergue, file = paste0("",out,"/assocclump.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", assocclumpmergue$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        }## CIERRE ANNOVAR NO CLUMP SI
      }## cierre clump Si
      else {
        t1 = filter(assocrelevant, assocrelevant$`     F_A` <= 0.05)
        t2 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.05 & assocrelevant$`     F_A` <= 0.1)
        t3 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.1 & assocrelevant$`     F_A` <= 0.2)
        t4 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.2 & assocrelevant$`     F_A` <= 0.3)
        t5 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.3 & assocrelevant$`     F_A` <= 0.4)
        t6 = filter(assocrelevant, assocrelevant$`     F_A` >= 0.4 )

        # Ordena un dataframe por una columna
        t1ordenado = arrange(t1, t1$`           P`)
        t2ordenado = arrange(t2, t2$`           P`)
        t3ordenado = arrange(t3, t3$`           P`)
        t4ordenado = arrange(t4, t4$`           P`)
        t5ordenado = arrange(t5, t5$`           P`)
        t6ordenado = arrange(t6, t6$`           P`)

        # Seleccionamos según filtrado
        t1seleccion = t1ordenado[1:5,]
        t2seleccion = t2ordenado[1:5,]
        t3seleccion = t3ordenado[1:20,]
        t4seleccion = t4ordenado[1:20,]
        t5seleccion = t5ordenado[1:20,]
        t6seleccion = t6ordenado[1:30,]

        seleccion = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                              t4seleccion, t5seleccion, t6seleccion)

        # Remove rows with all NAs
        seleccion <- seleccion[rowSums(is.na(seleccion)) != ncol(seleccion), ]

        ## Apertura clump no annovar si
        if (ANNOVAR) {# Preparamos para ANNOVAR
          # Preparamos para ANNOVAR
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR")
          seleccion$SNPA = seleccion$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            seleccion$SNP = gsub(paste0("",ii,":"), "", seleccion$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          assocanovar = data.frame(seleccion$CHR,seleccion$SNP,seleccion$SNP,
                                   seleccion$A2,seleccion$A1 , stringsAsFactors = F)

          # Format the file
          assocanovar$seleccion.CHR = as.numeric(assocanovar$seleccion.CHR)
          assocanovar$seleccion.SNP = as.numeric(assocanovar$seleccion.SNP)
          assocanovar$seleccion.SNP.1 = as.numeric(assocanovar$seleccion.SNP.1)
          assocanovar$seleccion.A2 = gsub("   ", "", assocanovar$seleccion.A2)
          assocanovar$seleccion.A1 = gsub("   ", "", assocanovar$seleccion.A1)

          # Write the file that is te ANNOVAR input
          write.table(assocanovar, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove quotes from the file
          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovar.txt"))


          # ANNOVAR command:
          # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))


          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove "SNV" from the file
          #system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)
          if(length(asevf) != 0){
            names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
          }
          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))
          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(seleccion, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux2 = unique(aux2)



            X = full_join(seleccion, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }



          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnncccc")


          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          # Filtra las variantes de la lista de genes
          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## cIERRE ANNOVAR SI CLUMP NO
        else{
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR")
          write.table(seleccion, file = paste0("",out,"/assocfilter.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", seleccion$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        }## CIERRE ANNOVAR NO CLUMP NO
      } ## Cierre clump no
    } ## Cierre covs no

    # Generación de epistasis (Parametrizar los threads)

    if (Wu){

      # Generation of epistasis with WU joint effects correction (Same as CASSI, does not test for set so we have to use PLINK) Joint-effects applies correction that there are 5 per cell of the contingency table.
      system(paste0("",plink," --bfile ",bfile," --fast-epistasis joint-effects --set-test --set ",out,"/epistasisrset.set --out ",out,"/episelecrjoint --threads ",core," --epi1 0.01"))
    }
    else{
      # Without Joint effects
      system(paste0("",plink," --bfile ",bfile," --fast-epistasis --set-test --set ",out,"/epistasisrset.set --out ",out,"/episelecr --threads ",core,""))
    }


    tx  = readLines(paste0("",out,"/episelecrjoint.epi.cc.summary"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/episelecrjoint.epi.cc.summary"),
                sep = " ", row.names = F, col.names = F, quote = F)

    tx  = readLines(paste0("",out,"/episelecrjoint.epi.cc"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/episelecrjoint.epi.cc"),
                sep = " ", row.names = F, col.names = F, quote = F)


    # Remove quotes and spaces
    #system(paste0("sed -i 's/\"//g' ",out,"/episelecrjoint.epi.cc.summary"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/episelecrjoint.epi.cc.summary"))
    #system(paste0("sed -i 's/\"//g' ",out,"/episelecrjoint.epi.cc"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/episelecrjoint.epi.cc"))

    tx  = readLines(paste0("",out,"/assocall.assoc"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/assocall.assoc"),
                sep = " ", row.names = F, col.names = F, quote = F)

    # Remove quotes and spaces
    #system(paste0("sed -i 's/\"//g' ",out,"/assocall.assoc"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocall.assoc"))

    # Read assoc data
    assocall100 = read_delim(paste0("",out,"/assocall.assoc"),
                             delim = " ", col_types = "ncncnncnnn")
    names(assocall100) = c("CHR" ,"SNP", "BP", "A1", "F_A", "F_U", "A2", "CHISQ", "P", "OR")

    # Read epistasis results
    episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                           delim = " ", col_types = "ncnnnnnc")
    names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
    episuc100 = mutate(episuc100, TBf = 0.05/episuc100$N_TOT)
    names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP", "TBf")

    # Read variants and annovar data
    assocrelevant100 = read_delim(paste0("",out,"/assocrelevant.txt"),
                                  delim = " ", col_types = "ncncnncnnn")
    asclumanovar100= read_delim(paste0("",out,"/aaclummerge.txt"),
                                delim = " ", col_types = "nnncnncnnncccccc")



    # Keep the epistasis SNP and mergue with assoc data

    aux = episuc100[8]
    names(aux) = "SNP"
    names(assocall100)[2] = "SNP"

    # Full_join by position
    X = inner_join(assocall100, aux, by ="SNP")

    X$SNPA = X$SNP

    # Replace the chromosome to keep the position in the SNP column
    for(ii in 22:1){
      X$SNP = gsub(paste0("",ii,":"), "", X$SNP)
    }

    # Format for ANNOVAR
    # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
    snpannovar = data.frame(X$CHR,X$SNP,X$SNP,X$A2,X$A1 , stringsAsFactors = F)

    # Format the file

    snpannovar$X.CHR = as.numeric(snpannovar$X.CHR)
    snpannovar$X.SNP = as.numeric(snpannovar$X.SNP)
    snpannovar$X.SNP.1 = as.numeric(snpannovar$X.SNP.1)
    snpannovar$X.A2 = gsub("   ", "", snpannovar$X.A2)
    snpannovar$X.A1 = gsub("   ", "", snpannovar$X.A1)

    # Write the file that is the ANNOVAR input
    write.table(snpannovar, file = paste0("",out,"/snpannovar.txt"),
                sep = " ", row.names = F, col.names = F, quote = F)


    # Remove quotes from the file
    #system(paste0("sed -i 's/\"//g' ",out,"/snpannovar.txt"))

    # ANNOVAR command:
    # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/
    system(paste0("",varpl," -out ",out,"/snpannovar -build ",build," ",out,"/snpannovar.txt ",db,""))


    tx  = readLines(paste0("",out,"/snpannovar.exonic_variant_function"))
    tx  = gsub(pattern = "SNV", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/snpannovar.exonic_variant_function"),
                sep = " ", row.names = F, col.names = F, quote = F)


    # Remove "SNV" from the file
    #system(paste0("sed -i 's/SNV//g' ",out,"/snpannovar.exonic_variant_function"))

    # Remove the spaces and tabs to generate only simple spaces throughout the file
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.exonic_variant_function"))

    # mequedo2
    #  Load in R ANNOVAR file
    asevf = read_delim(paste0("",out,"/snpannovar.exonic_variant_function"),
                       delim = " ", col_types = "cccccccc", col_names = F)
    if (length(asevf) != 0){
      names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
    }

    # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
    # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

    tx  = readLines(paste0("",out,"/snpannovar.variant_function"))
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/snpannovar.variant_function"),
                sep = " ", row.names = F, col.names = F, quote = F)



    # Same for the other file
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.variant_function"))
    asvf = read_delim(paste0("",out,"/snpannovar.variant_function"),
                      delim = " ", col_types = "ccccccc", col_names = F)
    names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

    #
    if (length(asevf) != 0){
      # Mergue data frames
      aux = asevf[c("tipo", "ini")]
      names(aux)[2] = "SNP"
      aux$SNP = as.numeric(aux$SNP)
      X$SNP = as.numeric(X$SNP)
      aux = unique(aux)

      # Full_join by position
      X = full_join(X, aux, by ="SNP")

      # Same for the other file
      aux2 = asvf[c("loc","gen","ini")]
      names(aux2)[3] = "SNP"
      aux2$SNP = as.numeric(aux2$SNP)
      aux2 = unique(aux2)


      X = full_join(X, aux2, by ="SNP")

      # Write the file with PLINK and ANNOVAR data
      write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)
    }else{
      aux2 = asvf[c("loc","gen","ini")]
      names(aux2)[3] = "SNP"
      aux2$SNP = as.numeric(aux2$SNP)
      X$SNP = as.numeric(X$SNP)
      aux2 = unique(aux2)



      X = full_join(X, aux2, by ="SNP")




      # Write the file with PLINK and ANNOVAR data
      write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)
    }

    tx  = readLines(paste0("",out,"/snpepistasisannovar.txt"))
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/snpepistasisannovar.txt"),
                sep = " ", row.names = F, col.names = F, quote = F)

    #system(paste0("sed -i 's/\"//g' ",out,"/snpepistasisannovar.txt"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpepistasisannovar.txt"))

    # mequedo3

    if (!is.null(covar)){
      mergueprimarisnp = asclumanovar100[c(2,5,11,12,14,15,16)]
      names(mergueprimarisnp) = c("SNP","F_A1","ORl1","Pl1","Tipo1","Loc1","Gen1")
    }else{
      mergueprimarisnp = asclumanovar100[c(2,5,10,9,12,13,14)]
      names(mergueprimarisnp) = c("SNP","F_A1","OR1","P1","Tipo1","Loc1","Gen1")
    }

    mergueprimarisnp$SNP = as.numeric(mergueprimarisnp$SNP)

    episuc100$SNPA = episuc100[2]

    # Replace the chromosome to keep the position in the SNP column
    for(ii in 22:1){
      episuc100$SNP = gsub(paste0("",ii,":"), "", episuc100$SNP)
    }

    episuc100$SNP = as.numeric(episuc100$SNP)


    # Full_join by position
    m1 = inner_join(episuc100, mergueprimarisnp, by ="SNP")

    # mequedo4
    if (!is.null(covar)){
      if (length(asevf) != 0){
        merguesecundarysnp = X[c(2,5,9,10,12,13,14)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Tipo2", "Loc2", "Gen2")
      } else{
        merguesecundarysnp = X[c(2,5,9,10,12,13)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Loc2", "Gen2")
      }
    }else{
      if (length(asevf) != 0){
        merguesecundarysnp = X[c(2,5,9,10,12,13,14)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","P2", "OR2", "Tipo2", "Loc2", "Gen2")
      } else{
        merguesecundarysnp = X[c(2,5,9,10,12,13)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","P2", "OR2", "Loc2", "Gen2")
      }
    }

    m1$SNPA2 = m1[8]

    # Replace the chromosome to keep the position in the SNP column
    for(ii in 22:1){
      m1$BEST_SNP = gsub(paste0("",ii,":"), "", m1$BEST_SNP)
    }

    m1$BEST_SNP = as.numeric(m1$BEST_SNP)
    merguesecundarysnp$BEST_SNP = as.numeric(merguesecundarysnp$BEST_SNP)
    merguesecundarysnp = unique(merguesecundarysnp)

    m2 = inner_join(m1, merguesecundarysnp, by ="BEST_SNP")

    if (!is.null(covar)){
      if (length(asevf) != 0){
        final = m2[c(10,6,17,16,23,11,12,13,14,15,18,19,20,21,22,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "TIPO1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "TIPO2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,16,17,3:15)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)
      }else{
        final = m2[c(10,6,17,16,22,11,12,13,15,18,19,20,21,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,14,15,3:13)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)

      }
    }else{
      if (length(asevf) != 0){
        final = m2[c(10,6,17,16,23,11,12,13,14,15,18,19,20,21,22,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "OR1", "P1", "TIPO1", "LOC1",
                         "F_A2", "P2", "OR2", "TIPO2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,16,17,3:15)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)
      }else{
        final = m2[c(10,6,17,16,22,11,12,13,15,18,19,20,21,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "OR1", "P1", "LOC1",
                         "F_A2", "P2", "OR2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,14,15,3:13)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)

      }
    }

  } # Genelist no primary


  # Genelist y primary
  if (!is.null(genelist) & !is.null(primarylist)){
    # Always ANNOVAR T
    ANNOVAR = T

    # Read genelist
    genes <- read_tsv(paste0("",genelist,""), col_names = F)
    names(genes) <- "GEN"

    # Read primarylist
    primary <- read_tsv(paste0("",primarylist,""), col_names = F)
    names(primary) <- "GEN"

    if (!is.null(covar)){

      # Logistic regresion
      system(paste0("",plink," --bfile ",bfile," --logistic --covar ",covar," --covar-number ",ncov," --out ",out,"/covariatesall --threads ",core,""))

      # Assoc analysis
      system(paste0("",plink," --bfile ",bfile," --assoc --out ",out,"/assocall --threads ",core,""))

      logistic = read_delim(paste0("",out,"/covariatesall.assoc.logistic"),
                            delim = " ", col_types = "ncnccnnnn")
      names(logistic) = c("CHR" ,"SNP", "BP", "A1", "TEST", "NMISS", "OR", "STAT", "P")
      logadd = filter(logistic, TEST == "       ADD")
      assoc = read_delim(paste0("",out,"/assocall.assoc"),
                         delim = " ", col_types = "ncncnncnnn")

      # Filter variants with treshold 10^-5
      assocrelevant = filter(logadd, logadd$P < AsT)

      write.table(assocrelevant , file = paste0("",out,"/assocrelevant.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)

      assocrelevant = read_delim(paste0("",out,"/assocrelevant.txt"),
                                 delim = " ", col_types = "ncnccnnnn")
      loginner = assocrelevant[c("SNP","OR","P")]
      names(loginner) = c("SNP" ,"ORl", "Pl")

      # Full_join by position assoc and logistic
      names(assoc)[2] = "SNP"
      assocrelevant = inner_join(assoc, loginner, by ="SNP")

      # Filter variants more frequently than 1% in our population in our case 4700 variants
      assocrelevant = filter(assocrelevant, assocrelevant$`     F_A` >= MAF | assocrelevant$`     F_U` >= MAF )

      if (clump){

        # Clumping analysis

        system(paste0("",plink," --bfile ",bfile," --clump ",out,"/assocall.assoc --out ",out,"/assocldclump --threads ",core,""))

        clumpdata = read_delim(paste0("",out,"/assocldclump.clumped"),
                               delim = " ", col_types = "nncnnnnnnnnc")

        # Mergue assocrelevant and clump
        aux = clumpdata[c(1, 2,3)]
        names(aux)[3] = "SNP"
        names(assocrelevant)[2] = "SNP"

        # Full_join by position, in our case 437 variants
        X = inner_join(assocrelevant, aux, by ="SNP")
        assocclumpmergue = X[,-c(13,14)]



        if (ANNOVAR){

          # Format for ANNOVAR
          names(assocclumpmergue) =c("CHR","SNP","BP","A1","F_A",
                                     "F_U","A2","CHISQ","P","OR","ORl","Pl")

          assocclumpmergue$SNPA = assocclumpmergue$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assocclumpmergue$SNP = gsub(paste0("",ii,":"),
                                        "", assocclumpmergue$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          asclumpano = data.frame(assocclumpmergue$CHR,assocclumpmergue$SNP,
                                  assocclumpmergue$SNP,assocclumpmergue$A2,
                                  assocclumpmergue$A1 , stringsAsFactors = F)

          # Format the file
          asclumpano$assocclumpmergue.CHR = as.numeric(asclumpano$assocclumpmergue.CHR)
          asclumpano$assocclumpmergue.SNP = as.numeric(asclumpano$assocclumpmergue.SNP)
          asclumpano$assocclumpmergue.SNP.1 = as.numeric(asclumpano$assocclumpmergue.SNP.1)
          asclumpano$assocclumpmergue.A2 = gsub("   ", "", asclumpano$assocclumpmergue.A2)
          asclumpano$assocclumpmergue.A1 = gsub("   ", "", asclumpano$assocclumpmergue.A1)

          # Write the file that is the ANNOVAR input
          write.table(asclumpano, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))



          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          if (length(asevf) != 0){
            names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
          }
          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))

          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

          # mequedo
          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(assocclumpmergue, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"), sep = " ",
                        row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux2 = unique(aux2)



            X = full_join(assocclumpmergue, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }

          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnnnncccc")

          # filterprimary
          filterprimary = data.frame()
          for(ii in 1:length(primary$GEN)){
            filterprimaryp <- filter(asclumanovar, asclumanovar$gen == primary$GEN[ii] | grepl( paste(primary$GEN[ii],"(", sep = "" ), asclumanovar$gen, fixed = TRUE) )
            filterprimary <- rbind(filterprimary,filterprimaryp)
          }

          ## Esto despues cuando dos listas
          # si no covs 11 y 12 si covs 13, 14

          if (nrow(filterprimary) > 100){
            # Filter variants by MAF
            t1 = filter(filterprimary, filterprimary$F_A <= 0.05)
            t2 = filter(filterprimary, filterprimary$F_A >= 0.05 & filterprimary$F_A <= 0.1)
            t3 = filter(filterprimary, filterprimary$F_A >= 0.1 & filterprimary$F_A <= 0.2)
            t4 = filter(filterprimary, filterprimary$F_A >= 0.2 & filterprimary$F_A <= 0.3)
            t5 = filter(filterprimary, filterprimary$F_A >= 0.3 & filterprimary$F_A <= 0.4)
            t6 = filter(filterprimary, filterprimary$F_A >= 0.4 )

            # Sort a dataframe by p.value column
            t1ordenado = arrange(t1, t1$Pl)
            t2ordenado = arrange(t2, t2$Pl)
            t3ordenado = arrange(t3, t3$Pl)
            t4ordenado = arrange(t4, t4$Pl)
            t5ordenado = arrange(t5, t5$Pl)
            t6ordenado = arrange(t6, t6$Pl)

            # We select certain variants according to p value
            t1seleccion = t1ordenado[1:5,]
            t2seleccion = t2ordenado[1:5,]
            t3seleccion = t3ordenado[1:20,]
            t4seleccion = t4ordenado[1:20,]
            t5seleccion = t5ordenado[1:20,]
            t6seleccion = t6ordenado[1:30,]

            assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                         t4seleccion, t5seleccion, t6seleccion)

            assocclumpmergue <- assocclumpmergue[rowSums(is.na(assocclumpmergue)) != ncol(assocclumpmergue), ]
            asclumanovar = assocclumpmergue
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          } else {
            asclumanovar = filterprimary
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }




          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          # Filtra las variantes de la lista de genes
          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        } ## ANNOVAR else de annovar no clump si
        else {

          names(assocclumpmergue) = c("CHR","SNP","BP","A1","F_A",
                                      "F_U","A2","CHISQ","P","OR",
                                      "ORl","Pl")

          write.table(assocclumpmergue, file = paste0("",out,"/assocclump.txt"),
                      sep = " ",row.names = F, col.names = F, quote = F)

          write.table(c("SET", assocclumpmergue$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ",row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        } ## cierre annovar no clump si
      } ## clump supuesto else de clump no (nuevo annovar)
      else {

        seleccion = assocrelevant

        if (ANNOVAR){
          # Preparamos para ANNOVAR
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR",
                               "ORl","Pl")
          seleccion$SNPA = seleccion$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            seleccion$SNP = gsub(paste0("",ii,":"), "", seleccion$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          assocanovar = data.frame(seleccion$CHR,seleccion$SNP,seleccion$SNP,
                                   seleccion$A2,seleccion$A1 , stringsAsFactors = F)

          # Format the file
          assocanovar$seleccion.CHR = as.numeric(assocanovar$seleccion.CHR)
          assocanovar$seleccion.SNP = as.numeric(assocanovar$seleccion.SNP)
          assocanovar$seleccion.SNP.1 = as.numeric(assocanovar$seleccion.SNP.1)
          assocanovar$seleccion.A2 = gsub("   ", "", assocanovar$seleccion.A2)
          assocanovar$seleccion.A1 = gsub("   ", "", assocanovar$seleccion.A1)

          # Write the file that is te ANNOVAR input
          write.table(assocanovar, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove quotes from the file
          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovar.txt"))


          # ANNOVAR command:
          # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))


          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove "SNV" from the file
          #system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)
          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))
          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(seleccion, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux2 = unique(aux2)



            X = full_join(seleccion, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }

          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnnnncccc")

            filterprimary = data.frame()
            for(ii in 1:length(primary$GEN)){
              filterprimaryp <- filter(asclumanovar, asclumanovar$gen == primary$GEN[ii] | grepl( paste(primary$GEN[ii],"(", sep = "" ), asclumanovar$gen, fixed = TRUE) )
              filterprimary <- rbind(filterprimary,filterprimaryp)
              }

          # pcf
          if (nrow(filterprimary) > 100){
            # Filter variants by MAF
            t1 = filter(filterprimary, filterprimary$F_A <= 0.05)
            t2 = filter(filterprimary, filterprimary$F_A >= 0.05 & filterprimary$F_A <= 0.1)
            t3 = filter(filterprimary, filterprimary$F_A >= 0.1 & filterprimary$F_A <= 0.2)
            t4 = filter(filterprimary, filterprimary$F_A >= 0.2 & filterprimary$F_A <= 0.3)
            t5 = filter(filterprimary, filterprimary$F_A >= 0.3 & filterprimary$F_A <= 0.4)
            t6 = filter(filterprimary, filterprimary$F_A >= 0.4 )

            # Sort a dataframe by p.value column
            t1ordenado = arrange(t1, t1$Pl)
            t2ordenado = arrange(t2, t2$Pl)
            t3ordenado = arrange(t3, t3$Pl)
            t4ordenado = arrange(t4, t4$Pl)
            t5ordenado = arrange(t5, t5$Pl)
            t6ordenado = arrange(t6, t6$Pl)

            # We select certain variants according to p value
            t1seleccion = t1ordenado[1:5,]
            t2seleccion = t2ordenado[1:5,]
            t3seleccion = t3ordenado[1:20,]
            t4seleccion = t4ordenado[1:20,]
            t5seleccion = t5ordenado[1:20,]
            t6seleccion = t6ordenado[1:30,]

            assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                         t4seleccion, t5seleccion, t6seleccion)

            assocclumpmergue <- assocclumpmergue[rowSums(is.na(assocclumpmergue)) != ncol(assocclumpmergue), ]
            asclumanovar = assocclumpmergue
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          } else {
            asclumanovar = filterprimary
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }


          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          # Filtra las variantes de la lista de genes
          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
          # clumpnoglt

        } ## Cierre annovar si clump No
        else{

          names(seleccion) = c("CHR","SNP","BP","A1","F_A","F_U","A2","CHISQ","P","OR")

          write.table(seleccion, file = paste0("",out,"/assocfilter.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", seleccion$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        } # Cierre de ANNOVAR no EN CLUMP NO
      } ## Cierre de clump no
    } ## Cierre covs si
    else{

      system(paste0("",plink," --bfile ",bfile," --assoc --out ",out,"/assocall --threads ",core,""))

      assoc = read_delim(paste0("",out,"/assocall.assoc"),
                         delim = " ", col_types = "ncncnncnnn")

      assocrelevant = filter(assoc, assoc$`           P` < AsT)

      write.table(assocrelevant , file = paste0("",out,"/assocrelevant.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)

      assocrelevant = read_delim(paste0("",out,"/assocrelevant.txt"),
                                 delim = " ", col_types = "ncncnncnnn")

      assocrelevant = filter(assocrelevant, assocrelevant$`     F_A` >= MAF)

      if (clump){
        # Clumping analysis

        system(paste0("",plink," --bfile ",bfile," --clump ",out,"/assocall.assoc --out ",out,"/assocldclump --threads ",core,""))

        clumpdata = read_delim(paste0("",out,"/assocldclump.clumped"),
                               delim = " ", col_types = "nncnnnnnnnnc")

        # Mergue assocrelevant and clump
        aux = clumpdata[c(1, 2,3)]
        names(aux)[3] = "SNP"
        names(assocrelevant)[2] = "SNP"

        # Full_join by position, in our case 437 variants
        X = inner_join(assocrelevant, aux, by ="SNP")
        assocclumpmergue = X[,-c(11,12)]



        if (ANNOVAR){

          # Format for ANNOVAR
          names(assocclumpmergue) =c("CHR","SNP","BP","A1","F_A",
                                     "F_U","A2","CHISQ","P","OR")

          assocclumpmergue$SNPA = assocclumpmergue$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assocclumpmergue$SNP = gsub(paste0("",ii,":"),
                                        "", assocclumpmergue$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          asclumpano = data.frame(assocclumpmergue$CHR,assocclumpmergue$SNP,
                                  assocclumpmergue$SNP,assocclumpmergue$A2,
                                  assocclumpmergue$A1 , stringsAsFactors = F)

          # Format the file
          asclumpano$assocclumpmergue.CHR = as.numeric(asclumpano$assocclumpmergue.CHR)
          asclumpano$assocclumpmergue.SNP = as.numeric(asclumpano$assocclumpmergue.SNP)
          asclumpano$assocclumpmergue.SNP.1 = as.numeric(asclumpano$assocclumpmergue.SNP.1)
          asclumpano$assocclumpmergue.A2 = gsub("   ", "", asclumpano$assocclumpmergue.A2)
          asclumpano$assocclumpmergue.A1 = gsub("   ", "", asclumpano$assocclumpmergue.A1)

          # Write the file that is the ANNOVAR input
          write.table(asclumpano, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))



          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          if (length(asevf) != 0){
            names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
          }
          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))

          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

          # mequedo
          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(assocclumpmergue, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"), sep = " ",
                        row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            assocclumpmergue$SNP = as.numeric(assocclumpmergue$SNP)
            aux2 = unique(aux2)



            X = full_join(assocclumpmergue, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }

          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnncccc")

          # filterprimary
          filterprimary = data.frame()
          for(ii in 1:length(primary$GEN)){
            filterprimaryp <- filter(asclumanovar, asclumanovar$gen == primary$GEN[ii] | grepl( paste(primary$GEN[ii],"(", sep = "" ), asclumanovar$gen, fixed = TRUE) )
            filterprimary <- rbind(filterprimary,filterprimaryp)
          }

          ## Esto despues cuando dos listas
          # si no covs 11 y 12 si covs 13, 14

          if (nrow(filterprimary) > 100){
            # Filter variants by MAF
            t1 = filter(filterprimary, filterprimary$F_A <= 0.05)
            t2 = filter(filterprimary, filterprimary$F_A >= 0.05 & filterprimary$F_A <= 0.1)
            t3 = filter(filterprimary, filterprimary$F_A >= 0.1 & filterprimary$F_A <= 0.2)
            t4 = filter(filterprimary, filterprimary$F_A >= 0.2 & filterprimary$F_A <= 0.3)
            t5 = filter(filterprimary, filterprimary$F_A >= 0.3 & filterprimary$F_A <= 0.4)
            t6 = filter(filterprimary, filterprimary$F_A >= 0.4 )

            # Sort a dataframe by p.value column
            t1ordenado = arrange(t1, t1$Pl)
            t2ordenado = arrange(t2, t2$Pl)
            t3ordenado = arrange(t3, t3$Pl)
            t4ordenado = arrange(t4, t4$Pl)
            t5ordenado = arrange(t5, t5$Pl)
            t6ordenado = arrange(t6, t6$Pl)

            # We select certain variants according to p value
            t1seleccion = t1ordenado[1:5,]
            t2seleccion = t2ordenado[1:5,]
            t3seleccion = t3ordenado[1:20,]
            t4seleccion = t4ordenado[1:20,]
            t5seleccion = t5ordenado[1:20,]
            t6seleccion = t6ordenado[1:30,]

            assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                         t4seleccion, t5seleccion, t6seleccion)

            assocclumpmergue <- assocclumpmergue[rowSums(is.na(assocclumpmergue)) != ncol(assocclumpmergue), ]
            asclumanovar = assocclumpmergue
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          } else {
            asclumanovar = filterprimary
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }




          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
        }## CIERRE ANNOVAR NO CLUMP SI
      }## cierre clump Si
      else {

        seleccion = assocrelevant

        if (ANNOVAR){
          # Preparamos para ANNOVAR
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR"
          )
          seleccion$SNPA = seleccion$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            seleccion$SNP = gsub(paste0("",ii,":"), "", seleccion$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          assocanovar = data.frame(seleccion$CHR,seleccion$SNP,seleccion$SNP,
                                   seleccion$A2,seleccion$A1 , stringsAsFactors = F)

          # Format the file
          assocanovar$seleccion.CHR = as.numeric(assocanovar$seleccion.CHR)
          assocanovar$seleccion.SNP = as.numeric(assocanovar$seleccion.SNP)
          assocanovar$seleccion.SNP.1 = as.numeric(assocanovar$seleccion.SNP.1)
          assocanovar$seleccion.A2 = gsub("   ", "", assocanovar$seleccion.A2)
          assocanovar$seleccion.A1 = gsub("   ", "", assocanovar$seleccion.A1)

          # Write the file that is te ANNOVAR input
          write.table(assocanovar, file = paste0("",out,"/assocannovar.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove quotes from the file
          #system(paste0("sed -i 's/\"//g' ",out,"/assocannovar.txt"))


          # ANNOVAR command:
          # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/

          system(paste0("",varpl," -out ",out,"/assocannovar -build ",build," ",out,"/assocannovar.txt ",db,""))


          tx  = readLines(paste0("",out,"/assocannovar.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)


          # Remove "SNV" from the file
          #system(paste0("sed -i 's/SNV//g' ",out,"/assocannovar.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.exonic_variant_function"))

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/assocannovar.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          if (length(asevf) != 0){
            names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
          }

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/assocannovar.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/assocannovar.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocannovar.variant_function"))
          asvf = read_delim(paste0("",out,"/assocannovar.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames

          if (length(asevf) != 0){
            aux = asevf[c("tipo", "ini")]
            names(aux)[2] = "SNP"
            aux$SNP = as.numeric(aux$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux = unique(aux)

            # Full_join by position
            X = full_join(seleccion, aux, by ="SNP")

            # Same for the other file
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            aux2 = unique(aux2)


            X = full_join(X, aux2, by ="SNP")

            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }else{
            aux2 = asvf[c("loc","gen","ini")]
            names(aux2)[3] = "SNP"
            aux2$SNP = as.numeric(aux2$SNP)
            seleccion$SNP = as.numeric(seleccion$SNP)
            aux2 = unique(aux2)



            X = full_join(seleccion, aux2, by ="SNP")




            # Write the file with PLINK and ANNOVAR data
            write.table(X, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }

          tx  = readLines(paste0("",out,"/aaclummerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/aaclummerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/aaclummerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/aaclummerge.txt"))

          asclumanovar = read_delim(paste0("",out,"/aaclummerge.txt"),
                                    delim = " ", col_types = "nnncnncnnncccc")

            filterprimary = data.frame()
            for(ii in 1:length(primary$GEN)){
              filterprimaryp <- filter(asclumanovar, asclumanovar$gen == primary$GEN[ii] | grepl( paste(primary$GEN[ii],"(", sep = "" ), asclumanovar$gen, fixed = TRUE) )
              filterprimary <- rbind(filterprimary,filterprimaryp)
              }

          # pcf
          if (nrow(filterprimary) > 100){
            # Filter variants by MAF
            t1 = filter(filterprimary, filterprimary$F_A <= 0.05)
            t2 = filter(filterprimary, filterprimary$F_A >= 0.05 & filterprimary$F_A <= 0.1)
            t3 = filter(filterprimary, filterprimary$F_A >= 0.1 & filterprimary$F_A <= 0.2)
            t4 = filter(filterprimary, filterprimary$F_A >= 0.2 & filterprimary$F_A <= 0.3)
            t5 = filter(filterprimary, filterprimary$F_A >= 0.3 & filterprimary$F_A <= 0.4)
            t6 = filter(filterprimary, filterprimary$F_A >= 0.4 )

            # Sort a dataframe by p.value column
            t1ordenado = arrange(t1, t1$P)
            t2ordenado = arrange(t2, t2$P)
            t3ordenado = arrange(t3, t3$P)
            t4ordenado = arrange(t4, t4$P)
            t5ordenado = arrange(t5, t5$P)
            t6ordenado = arrange(t6, t6$P)

            # We select certain variants according to p value
            t1seleccion = t1ordenado[1:5,]
            t2seleccion = t2ordenado[1:5,]
            t3seleccion = t3ordenado[1:20,]
            t4seleccion = t4ordenado[1:20,]
            t5seleccion = t5ordenado[1:20,]
            t6seleccion = t6ordenado[1:30,]

            assocclumpmergue = bind_rows(t1seleccion, t2seleccion, t3seleccion,
                                         t4seleccion, t5seleccion, t6seleccion)

            assocclumpmergue <- assocclumpmergue[rowSums(is.na(assocclumpmergue)) != ncol(assocclumpmergue), ]
            asclumanovar = assocclumpmergue
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          } else {
            asclumanovar = filterprimary
            write.table(asclumanovar, file = paste0("",out,"/aaclummerge.txt"),
                        sep = " ", row.names = F, col.names = T, quote = F)
          }


          # Format for ANNOVAR
          names(assoc) =c("CHR","SNP","BP","A1","F_A",
                          "F_U","A2","CHISQ","P","OR")

          assoc$SNPA = assoc$SNP

          # Replace the chromosome to keep the position in the SNP column
          for(ii in 22:1){
            assoc$SNP = gsub(paste0("",ii,":"),
                             "", assoc$SNP)
          }

          # Format for ANNOVAR
          # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
          allano = data.frame(assoc$CHR,assoc$SNP,
                              assoc$SNP,assoc$A2,
                              assoc$A1 , stringsAsFactors = F)

          # Format the file
          allano$assoc.CHR = as.numeric(allano$assoc.CHR)
          allano$assoc.SNP = as.numeric(allano$assoc.SNP)
          allano$assoc.SNP.1 = as.numeric(allano$assoc.SNP.1)
          allano$assoc.A2 = gsub("   ", "", allano$assoc.A2)
          allano$assoc.A1 = gsub("   ", "", allano$assoc.A1)

          # Write the file that is the ANNOVAR input
          write.table(allano, file = paste0("",out,"/allano.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)



          # ANNOVAR command:

          system(paste0("",varpl," -out ",out,"/allano -build ",build," ",out,"/allano.txt ",db,""))



          tx  = readLines(paste0("",out,"/allano.exonic_variant_function"))
          tx  = gsub(pattern = "SNV", replace = "", x = tx)
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.exonic_variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #  Load in R ANNOVAR file
          asevf = read_delim(paste0("",out,"/allano.exonic_variant_function"),
                             delim = " ", col_types = "cccccccc", col_names = F)

          # Remove "SNV" from the file
          # system(paste0("sed -i 's/SNV//g' ",out,"/allano.exonic_variant_function"))

          # Remove the spaces and tabs to generate only simple spaces throughout the file
          # system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.exonic_variant_function"))


          names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")

          # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
          # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

          tx  = readLines(paste0("",out,"/allano.variant_function"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allano.variant_function"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          # Same for the other file
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allano.variant_function"))

          asvf = read_delim(paste0("",out,"/allano.variant_function"),
                            delim = " ", col_types = "ccccccc", col_names = F)
          names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")


          # Mergue data frames
          aux = asevf[c("tipo", "ini")]
          names(aux)[2] = "SNP"
          aux$SNP = as.numeric(aux$SNP)
          assoc$SNP = as.numeric(assoc$SNP)

          # Full_join by position
          X = full_join(assoc, aux, by ="SNP")

          # Same for the other file
          aux2 = asvf[c("loc","gen","ini")]
          names(aux2)[3] = "SNP"
          aux2$SNP = as.numeric(aux2$SNP)


          X = full_join(X, aux2, by ="SNP")

          # Write the file with PLINK and ANNOVAR data
          write.table(X, file = paste0("",out,"/allanomerge.txt"), sep = " ",
                      row.names = F, col.names = T, quote = F)

          tx  = readLines(paste0("",out,"/allanomerge.txt"))
          tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
          write.table(tx, file = paste0("",out,"/allanomerge.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/allanomerge.txt"))
          #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/allanomerge.txt"))

          allassocano = read_delim(paste0("",out,"/allanomerge.txt"),
                                   delim = " ", col_types = "nnncnncnnncccc")

          # Filtra las variantes de la lista de genes
          filtervariants = data.frame()
          for(ii in 1:length(genes$GEN)){
            filtergene <- filter(allassocano, allassocano$gen == genes$GEN[ii] | grepl( paste(genes$GEN[ii],"(", sep = "" ), allassocano$gen, fixed = TRUE) )
            filtervariants <- rbind(filtervariants,filtergene)
          }

          names(filtervariants)[1] <- "CHR"

          write.table(c("SET", asclumanovar$SNPA, "END", "SET2", filtervariants$SNPA, "END"), file = paste0("",out,"/epistasisrset.set"), sep = " ",
                      row.names = F, col.names = F, quote = F)
          system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))
          # clumpnoglt
        } ## cIERRE ANNOVAR SI CLUMP NO
        else{
          names(seleccion) = c("CHR","SNP","BP","A1","F_A",
                               "F_U","A2","CHISQ","P","OR")
          write.table(seleccion, file = paste0("",out,"/assocfilter.txt"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          write.table(c("SET", seleccion$SNP, "END"),
                      file = paste0("",out,"/epistasisrset.set"),
                      sep = " ", row.names = F, col.names = F, quote = F)

          #system(paste0("sed -i 's/\"//g' ",out,"/epistasisrset.set"))

        }## CIERRE ANNOVAR NO CLUMP NO
      } ## Cierre clump no
    } ## Cierre covs no

    # Generación de epistasis (Parametrizar los threads)

    if (Wu){

      # Generation of epistasis with WU joint effects correction (Same as CASSI, does not test for set so we have to use PLINK) Joint-effects applies correction that there are 5 per cell of the contingency table.
      system(paste0("",plink," --bfile ",bfile," --fast-epistasis joint-effects --set-test --set ",out,"/epistasisrset.set --out ",out,"/episelecrjoint --threads ",core," --epi1 0.01"))
    }
    else{
      # Without Joint effects
      system(paste0("",plink," --bfile ",bfile," --fast-epistasis --set-test --set ",out,"/epistasisrset.set --out ",out,"/episelecr --threads ",core,""))
    }


    tx  = readLines(paste0("",out,"/episelecrjoint.epi.cc.summary"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/episelecrjoint.epi.cc.summary"),
                sep = " ", row.names = F, col.names = F, quote = F)

    tx  = readLines(paste0("",out,"/episelecrjoint.epi.cc"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/episelecrjoint.epi.cc"),
                sep = " ", row.names = F, col.names = F, quote = F)


    # Remove quotes and spaces
    #system(paste0("sed -i 's/\"//g' ",out,"/episelecrjoint.epi.cc.summary"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/episelecrjoint.epi.cc.summary"))
    #system(paste0("sed -i 's/\"//g' ",out,"/episelecrjoint.epi.cc"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/episelecrjoint.epi.cc"))

    tx  = readLines(paste0("",out,"/assocall.assoc"))
    tx  = gsub(pattern = "\"", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/assocall.assoc"),
                sep = " ", row.names = F, col.names = F, quote = F)

    # Remove quotes and spaces
    #system(paste0("sed -i 's/\"//g' ",out,"/assocall.assoc"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/assocall.assoc"))

    # Read assoc data
    assocall100 = read_delim(paste0("",out,"/assocall.assoc"),
                             delim = " ", col_types = "ncncnncnnn")
    names(assocall100) = c("CHR" ,"SNP", "BP", "A1", "F_A", "F_U", "A2", "CHISQ", "P", "OR")

    # Read epistasis results
    episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                           delim = " ", col_types = "ncnnnnnc")
    names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
    episuc100 = mutate(episuc100, TBf = 0.05/episuc100$N_TOT)
    names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP", "TBf")

    # Read variants and annovar data

    if (!is.null(covar)){
      assocrelevant100 = read_delim(paste0("",out,"/assocrelevant.txt"),
                                    delim = " ", col_types = "ncncnncnnn")
      asclumanovar100= read_delim(paste0("",out,"/aaclummerge.txt"),
                                  delim = " ", col_types = "nnncnncnnncccccc")
    }else{
      assocrelevant100 = read_delim(paste0("",out,"/assocrelevant.txt"),
                                    delim = " ", col_types = "ncncnncnnn")
      asclumanovar100= read_delim(paste0("",out,"/aaclummerge.txt"),
                                  delim = " ", col_types = "nnncnncnnncccc")
    }


    # Keep the epistasis SNP and mergue with assoc data

    aux = episuc100[8]
    names(aux) = "SNP"
    names(assocall100)[2] = "SNP"

    # Full_join by position
    X = inner_join(assocall100, aux, by ="SNP")

    X$SNPA = X$SNP

    # Replace the chromosome to keep the position in the SNP column
    for(ii in 22:1){
      X$SNP = gsub(paste0("",ii,":"), "", X$SNP)
    }

    # Format for ANNOVAR
    # We have to specify chr, start position, end position, reference allele (major), alternative allele (minor)
    snpannovar = data.frame(X$CHR,X$SNP,X$SNP,X$A2,X$A1 , stringsAsFactors = F)

    # Format the file

    snpannovar$X.CHR = as.numeric(snpannovar$X.CHR)
    snpannovar$X.SNP = as.numeric(snpannovar$X.SNP)
    snpannovar$X.SNP.1 = as.numeric(snpannovar$X.SNP.1)
    snpannovar$X.A2 = gsub("   ", "", snpannovar$X.A2)
    snpannovar$X.A1 = gsub("   ", "", snpannovar$X.A1)

    # Write the file that is the ANNOVAR input
    write.table(snpannovar, file = paste0("",out,"/snpannovar.txt"),
                sep = " ", row.names = F, col.names = F, quote = F)


    # Remove quotes from the file
    #system(paste0("sed -i 's/\"//g' ",out,"/snpannovar.txt"))

    # ANNOVAR command:
    # annotate_variation.pl -out c1controlesannovar -build hg19 chr1controlesannovar.txt Doctorado/Annovar/annovar.latest/annovar/humandb/
    system(paste0("",varpl," -out ",out,"/snpannovar -build ",build," ",out,"/snpannovar.txt ",db,""))


    tx  = readLines(paste0("",out,"/snpannovar.exonic_variant_function"))
    tx  = gsub(pattern = "SNV", replace = "", x = tx)
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/snpannovar.exonic_variant_function"),
                sep = " ", row.names = F, col.names = F, quote = F)


    # Remove "SNV" from the file
    #system(paste0("sed -i 's/SNV//g' ",out,"/snpannovar.exonic_variant_function"))

    # Remove the spaces and tabs to generate only simple spaces throughout the file
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.exonic_variant_function"))

    # mequedo2
    #  Load in R ANNOVAR file
    asevf = read_delim(paste0("",out,"/snpannovar.exonic_variant_function"),
                       delim = " ", col_types = "cccccccc", col_names = F)
    if (length(asevf) != 0){
      names(asevf) = c("linea","tipo","gen","chr","ini","fin","ref","alt")
    }

    # Possible values in field "tipo" include: nonsynonymous, synonymous, frameshift insertion,
    # frameshift deletion, nonframeshift insertion, nonframeshift deletion, frameshift block substitution, nonframshift block substitution

    tx  = readLines(paste0("",out,"/snpannovar.variant_function"))
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/snpannovar.variant_function"),
                sep = " ", row.names = F, col.names = F, quote = F)



    # Same for the other file
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpannovar.variant_function"))
    asvf = read_delim(paste0("",out,"/snpannovar.variant_function"),
                      delim = " ", col_types = "ccccccc", col_names = F)
    names(asvf) = c("loc","gen","chr","ini","fin","ref","alt")

    #
    if (length(asevf) != 0){
      # Mergue data frames
      aux = asevf[c("tipo", "ini")]
      names(aux)[2] = "SNP"
      aux$SNP = as.numeric(aux$SNP)
      X$SNP = as.numeric(X$SNP)
      aux = unique(aux)

      # Full_join by position
      X = full_join(X, aux, by ="SNP")

      # Same for the other file
      aux2 = asvf[c("loc","gen","ini")]
      names(aux2)[3] = "SNP"
      aux2$SNP = as.numeric(aux2$SNP)
      aux2 = unique(aux2)


      X = full_join(X, aux2, by ="SNP")

      # Write the file with PLINK and ANNOVAR data
      write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)
    }else{
      aux2 = asvf[c("loc","gen","ini")]
      names(aux2)[3] = "SNP"
      aux2$SNP = as.numeric(aux2$SNP)
      X$SNP = as.numeric(X$SNP)
      aux2 = unique(aux2)



      X = full_join(X, aux2, by ="SNP")




      # Write the file with PLINK and ANNOVAR data
      write.table(X, file = paste0("",out,"/snpepistasisannovar.txt"),
                  sep = " ", row.names = F, col.names = T, quote = F)
    }

    tx  = readLines(paste0("",out,"/snpepistasisannovar.txt"))
    tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
    write.table(tx, file = paste0("",out,"/snpepistasisannovar.txt"),
                sep = " ", row.names = F, col.names = F, quote = F)

    #system(paste0("sed -i 's/\"//g' ",out,"/snpepistasisannovar.txt"))
    #system(paste0("sed -i 's/[[:space:]]\\+/ /g' ",out,"/snpepistasisannovar.txt"))

    # mequedo3

    if (!is.null(covar)){
      mergueprimarisnp = asclumanovar100[c(2,5,11,12,14,15,16)]
      names(mergueprimarisnp) = c("SNP","F_A1","ORl1","Pl1","Tipo1","Loc1","Gen1")
    }else{
      mergueprimarisnp = asclumanovar100[c(2,5,10,9,12,13,14)]
      names(mergueprimarisnp) = c("SNP","F_A1","OR1","P1","Tipo1","Loc1","Gen1")
    }

    mergueprimarisnp$SNP = as.numeric(mergueprimarisnp$SNP)

    episuc100$SNPA = episuc100[2]

    # Replace the chromosome to keep the position in the SNP column
    for(ii in 22:1){
      episuc100$SNP = gsub(paste0("",ii,":"), "", episuc100$SNP)
    }

    episuc100$SNP = as.numeric(episuc100$SNP)


    # Full_join by position
    m1 = inner_join(episuc100, mergueprimarisnp, by ="SNP")

    # mequedo4
    if (!is.null(covar)){
      if (length(asevf) != 0){
        merguesecundarysnp = X[c(2,5,9,10,12,13,14)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Tipo2", "Loc2", "Gen2")
      } else{
        merguesecundarysnp = X[c(2,5,9,10,12,13)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","Pl2", "ORl2", "Loc2", "Gen2")
      }
    }else{
      if (length(asevf) != 0){
        merguesecundarysnp = X[c(2,5,9,10,12,13,14)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","P2", "OR2", "Tipo2", "Loc2", "Gen2")
      } else{
        merguesecundarysnp = X[c(2,5,9,10,12,13)]

        names(merguesecundarysnp)= c("BEST_SNP","F_A2","P2", "OR2", "Loc2", "Gen2")
      }
    }

    m1$SNPA2 = m1[8]

    # Replace the chromosome to keep the position in the SNP column
    for(ii in 22:1){
      m1$BEST_SNP = gsub(paste0("",ii,":"), "", m1$BEST_SNP)
    }

    m1$BEST_SNP = as.numeric(m1$BEST_SNP)
    merguesecundarysnp$BEST_SNP = as.numeric(merguesecundarysnp$BEST_SNP)
    merguesecundarysnp = unique(merguesecundarysnp)

    m2 = inner_join(m1, merguesecundarysnp, by ="BEST_SNP")

    if (!is.null(covar)){
      if (length(asevf) != 0){
        final = m2[c(10,6,17,16,23,11,12,13,14,15,18,19,20,21,22,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "TIPO1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "TIPO2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,16,17,3:15)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)
      }else{
        final = m2[c(10,6,17,16,22,11,12,13,15,18,19,20,21,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "ORl1", "Pl1", "LOC1",
                         "F_A2", "Pl2", "ORl2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,14,15,3:13)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)

      }
    }else{
      if (length(asevf) != 0){
        final = m2[c(10,6,17,16,23,11,12,13,14,15,18,19,20,21,22,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "OR1", "P1", "TIPO1", "LOC1",
                         "F_A2", "P2", "OR2", "TIPO2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,16,17,3:15)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)
      }else{
        final = m2[c(10,6,17,16,22,11,12,13,15,18,19,20,21,9)]

        names(final) = c("SNP", "CHISQ", "SNP2", "GEN1", "GEN2",
                         "F_A1", "OR1", "P1", "LOC1",
                         "F_A2", "P2", "OR2", "LOC2", "TBf")

        final2 = as.matrix(final)

        final3 = as.data.frame(final2)

        # To keep p.value
        episuc100 = read_delim(paste0("",out,"/episelecrjoint.epi.cc.summary"),
                               delim = " ", col_types = "ncnnnnnc")
        names(episuc100) = c("CHR", "SNP", "N_SIG", "N_TOT", "PROP", "BEST_CHISQ", "BEST_CHR", "BEST_SNP")
        episet = episuc100
        epirep = episet[c(2,8)]
        epipvalue = read_delim(paste0("",out,"/episelecrjoint.epi.cc"),
                               delim = " ", col_types = "ncncnn")
        names(epipvalue) = c("CHR1", "SNP1", "CHR2", "SNP2", "STAT", "P")

        quiero = filter(epipvalue, epipvalue$SNP1 %in% epirep$SNP & epipvalue$SNP2 %in% epirep$BEST_SNP)
        quiero = mutate(quiero, deseo = paste(SNP1,SNP2))
        episet = mutate(episet, deseo = paste(SNP,BEST_SNP))
        qf = filter(quiero, quiero$deseo %in% episet$deseo)
        qf = qf[c(2,6)]
        names(qf) = c("SNP", "Pepi")

        m5 = inner_join(final3, qf, by ="SNP")

        epiinform = m5[c(1,2,14,15,3:13)]
        write.table(epiinform, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = T, quote = F)

        tx  = readLines(paste0("",out,"//epiinform.txt"))
        tx  = gsub(pattern = "[[:space:]]", replace = " ", x = tx)
        write.table(tx, file = paste0("",out,"/epiinform.txt"),
                    sep = " ", row.names = F, col.names = F, quote = F)

      }
    }

  } # Genelist y primary


} # Fin de función
